// @v-c/utils/dist/dom/css.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/dom/css.ts
var css_exports = {};
__export(css_exports, {
  getClientSize: () => getClientSize,
  getCss: () => getCss,
  getDocSize: () => getDocSize,
  getOffset: () => getOffset,
  getOuterHeight: () => getOuterHeight,
  getOuterWidth: () => getOuterWidth,
  getScroll: () => getScroll,
  setCss: () => setCss
});
module.exports = __toCommonJS(css_exports);
var PIXEL_PATTERN = /margin|padding|width|height|max|min|offset/;
var removePixel = {
  left: true,
  top: true
};
var floatMap = {
  cssFloat: 1,
  styleFloat: 1,
  float: 1
};
function getComputedStyle(node) {
  return node.nodeType === 1 ? node.ownerDocument.defaultView.getComputedStyle(node, null) : {};
}
function getStyleValue(node, type, value) {
  type = type.toLowerCase();
  if (value === "auto") {
    if (type === "height")
      return node.offsetHeight;
    if (type === "width")
      return node.offsetWidth;
  }
  if (!(type in removePixel))
    removePixel[type] = PIXEL_PATTERN.test(type);
  return removePixel[type] ? parseFloat(value) || 0 : value;
}
function getCss(node, name) {
  const length = arguments.length;
  const style = getComputedStyle(node);
  name = floatMap[name] ? "cssFloat" in node.style ? "cssFloat" : "styleFloat" : name;
  return length === 1 ? style : getStyleValue(node, name, style[name] || node.style[name]);
}
function setCss(node, name, value) {
  const length = arguments.length;
  name = floatMap[name] ? "cssFloat" in node.style ? "cssFloat" : "styleFloat" : name;
  if (length === 3) {
    if (typeof value === "number" && PIXEL_PATTERN.test(name))
      value = `${value}px`;
    node.style[name] = value;
    return value;
  }
  for (const x in name) {
    if (name.hasOwnProperty(x))
      setCss(node, x, name[x]);
  }
  return getComputedStyle(node);
}
function getOuterWidth(el) {
  if (el === document.body)
    return document.documentElement.clientWidth;
  return el.offsetWidth;
}
function getOuterHeight(el) {
  if (el === document.body)
    return window.innerHeight || document.documentElement.clientHeight;
  return el.offsetHeight;
}
function getDocSize() {
  const width = Math.max(document.documentElement.scrollWidth, document.body.scrollWidth);
  const height = Math.max(document.documentElement.scrollHeight, document.body.scrollHeight);
  return {
    width,
    height
  };
}
function getClientSize() {
  const width = document.documentElement.clientWidth;
  const height = window.innerHeight || document.documentElement.clientHeight;
  return {
    width,
    height
  };
}
function getScroll() {
  return {
    scrollLeft: Math.max(document.documentElement.scrollLeft, document.body.scrollLeft),
    scrollTop: Math.max(document.documentElement.scrollTop, document.body.scrollTop)
  };
}
function getOffset(node) {
  const box = node.getBoundingClientRect();
  const docElem = document.documentElement;
  return {
    left: box.left + (window.pageXOffset || docElem.scrollLeft) - (docElem.clientLeft || document.body.clientLeft || 0),
    top: box.top + (window.pageYOffset || docElem.scrollTop) - (docElem.clientTop || document.body.clientTop || 0)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getClientSize,
  getCss,
  getDocSize,
  getOffset,
  getOuterHeight,
  getOuterWidth,
  getScroll,
  setCss
});
