// @v-c/utils/dist/dom/focus.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/dom/focus.ts
var focus_exports = {};
__export(focus_exports, {
  backLastFocusNode: () => backLastFocusNode,
  clearLastFocusNode: () => clearLastFocusNode,
  getFocusNodeList: () => getFocusNodeList,
  limitTabRange: () => limitTabRange,
  saveLastFocusNode: () => saveLastFocusNode
});
module.exports = __toCommonJS(focus_exports);

// src/dom/is-visible.ts
var is_visible_default = (element) => {
  if (!element)
    return false;
  if (element instanceof Element) {
    if (element.offsetParent)
      return true;
    if (element.getBBox) {
      const { width, height } = element.getBBox();
      if (width || height)
        return true;
    }
    if (element.getBoundingClientRect) {
      const { width, height } = element.getBoundingClientRect();
      if (width || height)
        return true;
    }
  }
  return false;
};

// src/dom/focus.ts
function focusable(node, includePositive = false) {
  if (is_visible_default(node)) {
    const nodeName = node.nodeName.toLowerCase();
    const isFocusableElement = ["input", "select", "textarea", "button"].includes(nodeName) || node.isContentEditable || nodeName === "a" && !!node.getAttribute("href");
    const tabIndexAttr = node.getAttribute("tabindex");
    const tabIndexNum = Number(tabIndexAttr);
    let tabIndex = null;
    if (tabIndexAttr && !Number.isNaN(tabIndexNum))
      tabIndex = tabIndexNum;
    else if (isFocusableElement && tabIndex === null)
      tabIndex = 0;
    if (isFocusableElement && node.disabled)
      tabIndex = null;
    return tabIndex !== null && (tabIndex >= 0 || includePositive && tabIndex < 0);
  }
  return false;
}
function getFocusNodeList(node, includePositive = false) {
  const res = [...node.querySelectorAll("*")].filter((child) => {
    return focusable(child, includePositive);
  });
  if (focusable(node, includePositive))
    res.unshift(node);
  return res;
}
var lastFocusElement = null;
function saveLastFocusNode() {
  lastFocusElement = document.activeElement;
}
function clearLastFocusNode() {
  lastFocusElement = null;
}
function backLastFocusNode() {
  if (lastFocusElement) {
    try {
      lastFocusElement.focus();
    } catch (e) {
    }
  }
}
function limitTabRange(node, e) {
  if (e.keyCode === 9) {
    const tabNodeList = getFocusNodeList(node);
    const lastTabNode = tabNodeList[e.shiftKey ? 0 : tabNodeList.length - 1];
    const leavingTab = lastTabNode === document.activeElement || node === document.activeElement;
    if (leavingTab) {
      const target = tabNodeList[e.shiftKey ? tabNodeList.length - 1 : 0];
      target.focus();
      e.preventDefault();
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  backLastFocusNode,
  clearLastFocusNode,
  getFocusNodeList,
  limitTabRange,
  saveLastFocusNode
});
