// @v-c/utils/dist/dom/index.js
import "../chunk-AX4OQNC5.js";
import {
  backLastFocusNode,
  clearLastFocusNode,
  getFocusNodeList,
  limitTabRange,
  saveLastFocusNode
} from "../chunk-3U7JRFHY.js";
import {
  isValidElement
} from "../chunk-OY5YTCGF.js";
import {
  is_visible_default
} from "../chunk-6FJXPUUP.js";
import {
  isStyleSupport
} from "../chunk-N4Q4R4CM.js";
import {
  styleNames
} from "../chunk-R6HVVLW2.js";
import {
  animation,
  transition
} from "../chunk-YHR5IYSS.js";
import {
  addClass,
  hasClass,
  removeClass
} from "../chunk-CIJVO62H.js";
import {
  getClientSize,
  getCss,
  getDocSize,
  getOffset,
  getOuterHeight,
  getOuterWidth,
  getScroll,
  setCss
} from "../chunk-V5Z5Q6SK.js";
import {
  clearContainerCache,
  injectCSS,
  removeCSS,
  updateCSS
} from "../chunk-AKT5ONZF.js";
import "../chunk-QJ2CZK2Z.js";
import {
  contains
} from "../chunk-HN3FVSOV.js";
import {
  find_dom_node_default
} from "../chunk-V3URHDQ7.js";
import "../chunk-AHWDZYGD.js";
export {
  addClass,
  animation,
  backLastFocusNode,
  clearContainerCache,
  clearLastFocusNode,
  contains,
  find_dom_node_default as findDOMNode,
  find_dom_node_default as findDomNode,
  getClientSize,
  getCss,
  getDocSize,
  getFocusNodeList,
  getOffset,
  getOuterHeight,
  getOuterWidth,
  getScroll,
  hasClass,
  injectCSS,
  isStyleSupport,
  isValidElement,
  is_visible_default as isVisible,
  limitTabRange,
  removeCSS,
  removeClass,
  saveLastFocusNode,
  setCss,
  styleNames,
  transition,
  updateCSS
};
