// @v-c/utils/dist/dom/support.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/dom/support.ts
var support_exports = {};
__export(support_exports, {
  animation: () => animation,
  transition: () => transition
});
module.exports = __toCommonJS(support_exports);

// src/lib/can-use-dom.ts
function canUseDom() {
  return !!(typeof window !== "undefined" && window.document && window.document.createElement);
}
var can_use_dom_default = canUseDom;

// src/dom/support.ts
var animationEndEventNames = {
  WebkitAnimation: "webkitAnimationEnd",
  OAnimation: "oAnimationEnd",
  animation: "animationend"
};
var transitionEventNames = {
  WebkitTransition: "webkitTransitionEnd",
  OTransition: "oTransitionEnd",
  transition: "transitionend"
};
function supportEnd(names) {
  const el = document.createElement("div");
  for (const name in names) {
    if (names.hasOwnProperty(name) && el.style[name] !== void 0) {
      return {
        end: names[name]
      };
    }
  }
  return false;
}
var animation = can_use_dom_default() && supportEnd(animationEndEventNames);
var transition = can_use_dom_default() && supportEnd(transitionEventNames);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  animation,
  transition
});
