// @v-c/utils/dist/base.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/base.ts
var base_exports = {};
__export(base_exports, {
  isArray: () => isArray,
  isBaseType: () => isBaseType,
  isBoolean: () => isBoolean,
  isDate: () => isDate,
  isDef: () => isDef,
  isFunction: () => isFunction,
  isNull: () => isNull,
  isNumber: () => isNumber,
  isObject: () => isObject,
  isString: () => isString,
  isSymbol: () => isSymbol,
  isUndefined: () => isUndefined,
  isUrl: () => isUrl,
  omit: () => omit,
  pick: () => pick,
  toArray: () => toArray,
  tuple: () => tuple,
  tupleArr: () => tupleArr,
  tupleNum: () => tupleNum,
  tupleNumArr: () => tupleNumArr
});
module.exports = __toCommonJS(base_exports);
var isString = (val) => typeof val === "string";
var isNumber = (val) => typeof val === "number";
var isBoolean = (val) => typeof val === "boolean";
var isFunction = (val) => typeof val === "function";
var isArray = (val) => Array.isArray(val);
var isUndefined = (val) => typeof val === "undefined";
var isDef = (val) => typeof val !== "undefined";
var isNull = (val) => val === null;
var isSymbol = (val) => typeof val === "symbol";
var isObject = (val) => Object.prototype.toString.call(val) === "[object Object]";
var isDate = (val) => Object.prototype.toString.call(val) === "[object Date]";
var isBaseType = (val) => {
  return isString(val) || isNumber(val) || isBoolean(val) || isSymbol(val) || isUndefined(val) || isNull(val);
};
var tuple = (...args) => args;
var tupleNum = (...args) => args;
function pick(obj, keys = [], rest) {
  const newObj = {};
  const names = Object.getOwnPropertyNames(obj);
  names.forEach((name) => {
    if (keys.includes(name))
      newObj[name] = obj[name];
  });
  return Object.assign(newObj, rest);
}
function omit(object, keys = [], rest) {
  if (object === null || object === void 0)
    return rest || object;
  const omitedObject = {};
  const originalKeys = Object.getOwnPropertyNames(object);
  originalKeys.forEach((originalKey) => {
    if (!keys.includes(originalKey))
      omitedObject[originalKey] = object[originalKey];
  });
  return Object.assign(omitedObject, rest);
}
var isUrl = (val) => {
  const reg = /^(https?|mailto|tel|file):/;
  return reg.test(val);
};
var tupleArr = (...args) => args;
var tupleNumArr = (...args) => args;
var toArray = (val) => {
  return isArray(val) ? val : [val];
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isArray,
  isBaseType,
  isBoolean,
  isDate,
  isDef,
  isFunction,
  isNull,
  isNumber,
  isObject,
  isString,
  isSymbol,
  isUndefined,
  isUrl,
  omit,
  pick,
  toArray,
  tuple,
  tupleArr,
  tupleNum,
  tupleNumArr
});
