// @v-c/utils/dist/chunk-3U7JRFHY.js
import {
  is_visible_default
} from "./chunk-6FJXPUUP.js";

// src/dom/focus.ts
function focusable(node, includePositive = false) {
  if (is_visible_default(node)) {
    const nodeName = node.nodeName.toLowerCase();
    const isFocusableElement = ["input", "select", "textarea", "button"].includes(nodeName) || node.isContentEditable || nodeName === "a" && !!node.getAttribute("href");
    const tabIndexAttr = node.getAttribute("tabindex");
    const tabIndexNum = Number(tabIndexAttr);
    let tabIndex = null;
    if (tabIndexAttr && !Number.isNaN(tabIndexNum))
      tabIndex = tabIndexNum;
    else if (isFocusableElement && tabIndex === null)
      tabIndex = 0;
    if (isFocusableElement && node.disabled)
      tabIndex = null;
    return tabIndex !== null && (tabIndex >= 0 || includePositive && tabIndex < 0);
  }
  return false;
}
function getFocusNodeList(node, includePositive = false) {
  const res = [...node.querySelectorAll("*")].filter((child) => {
    return focusable(child, includePositive);
  });
  if (focusable(node, includePositive))
    res.unshift(node);
  return res;
}
var lastFocusElement = null;
function saveLastFocusNode() {
  lastFocusElement = document.activeElement;
}
function clearLastFocusNode() {
  lastFocusElement = null;
}
function backLastFocusNode() {
  if (lastFocusElement) {
    try {
      lastFocusElement.focus();
    } catch (e) {
    }
  }
}
function limitTabRange(node, e) {
  if (e.keyCode === 9) {
    const tabNodeList = getFocusNodeList(node);
    const lastTabNode = tabNodeList[e.shiftKey ? 0 : tabNodeList.length - 1];
    const leavingTab = lastTabNode === document.activeElement || node === document.activeElement;
    if (leavingTab) {
      const target = tabNodeList[e.shiftKey ? tabNodeList.length - 1 : 0];
      target.focus();
      e.preventDefault();
    }
  }
}

export {
  getFocusNodeList,
  saveLastFocusNode,
  clearLastFocusNode,
  backLastFocusNode,
  limitTabRange
};
