// @v-c/utils/dist/chunk-AHWDZYGD.js
// src/base.ts
var isString = (val) => typeof val === "string";
var isNumber = (val) => typeof val === "number";
var isBoolean = (val) => typeof val === "boolean";
var isFunction = (val) => typeof val === "function";
var isArray = (val) => Array.isArray(val);
var isUndefined = (val) => typeof val === "undefined";
var isDef = (val) => typeof val !== "undefined";
var isNull = (val) => val === null;
var isSymbol = (val) => typeof val === "symbol";
var isObject = (val) => Object.prototype.toString.call(val) === "[object Object]";
var isDate = (val) => Object.prototype.toString.call(val) === "[object Date]";
var isBaseType = (val) => {
  return isString(val) || isNumber(val) || isBoolean(val) || isSymbol(val) || isUndefined(val) || isNull(val);
};
var tuple = (...args) => args;
var tupleNum = (...args) => args;
function pick(obj, keys = [], rest) {
  const newObj = {};
  const names = Object.getOwnPropertyNames(obj);
  names.forEach((name) => {
    if (keys.includes(name))
      newObj[name] = obj[name];
  });
  return Object.assign(newObj, rest);
}
function omit(object, keys = [], rest) {
  if (object === null || object === void 0)
    return rest || object;
  const omitedObject = {};
  const originalKeys = Object.getOwnPropertyNames(object);
  originalKeys.forEach((originalKey) => {
    if (!keys.includes(originalKey))
      omitedObject[originalKey] = object[originalKey];
  });
  return Object.assign(omitedObject, rest);
}
var isUrl = (val) => {
  const reg = /^(https?|mailto|tel|file):/;
  return reg.test(val);
};
var tupleArr = (...args) => args;
var tupleNumArr = (...args) => args;
var toArray = (val) => {
  return isArray(val) ? val : [val];
};

export {
  isString,
  isNumber,
  isBoolean,
  isFunction,
  isArray,
  isUndefined,
  isDef,
  isNull,
  isSymbol,
  isObject,
  isDate,
  isBaseType,
  tuple,
  tupleNum,
  pick,
  omit,
  isUrl,
  tupleArr,
  tupleNumArr,
  toArray
};
