// @v-c/utils/dist/chunk-BJR6PAYP.js
import {
  isFunction
} from "./chunk-AHWDZYGD.js";

// src/vnode.ts
import { Comment, Fragment, Text, vShow } from "vue";
var getSlotsProps = (slots, props, key, ...args) => {
  if (key in slots && isFunction(slots[key])) {
    const slot = slots[key];
    return slot?.(...args);
  }
  if (key in props) {
    const prop = props[key];
    if (isFunction(prop))
      return prop(...args);
    return prop;
  }
  return null;
};
var hasSlotProps = (slots, props, key) => {
  if (key in slots && slots[key])
    return true;
  return !!(key in props && props[key]);
};
function isEmptyElement(c) {
  return c && (c.type === Comment || c.type === Fragment && c.children.length === 0 || c.type === Text && c.children.trim() === "");
}
function filterEmpty(children = []) {
  const res = [];
  children.forEach((child) => {
    if (Array.isArray(child))
      res.push(...child);
    else if (child?.type === Fragment)
      res.push(...filterEmpty(child.children));
    else
      res.push(child);
  });
  return res.filter((c) => !isEmptyElement(c));
}
function isNodeVShowFalse(vNode) {
  const showDir = vNode.dirs?.find(({ dir }) => dir === vShow);
  return !!(showDir && showDir.value === false);
}
function getSlot(instance, slotName = "default", fallback = []) {
  const slots = instance?.$slots || instance.slots || {};
  const slot = slots[slotName];
  if (slot === void 0)
    return fallback;
  return slot();
}

export {
  getSlotsProps,
  hasSlotProps,
  isEmptyElement,
  filterEmpty,
  isNodeVShowFalse,
  getSlot
};
