// @v-c/utils/dist/chunk-ST7CAO4R.js
// src/type-util.ts
function stringType(defaultVal) {
  return {
    type: String,
    default: defaultVal
  };
}
function numberType(defaultVal) {
  return {
    type: Number,
    default: defaultVal
  };
}
function booleanType(defaultVal) {
  return {
    type: Boolean,
    default: defaultVal
  };
}
function arrayType(defaultVal) {
  return {
    type: Array,
    default: defaultVal
  };
}
function objectType(defaultVal) {
  return {
    type: Object,
    default: defaultVal
  };
}
function functionType(defaultVal) {
  return {
    type: Function,
    default: defaultVal
  };
}
function anyType(defaultVal, required) {
  const type = { validator: () => true, default: defaultVal };
  return required ? type : type;
}
function vNodeType() {
  return { validator: () => true };
}
function someType(types, defaultVal) {
  return types ? { type: types, default: defaultVal } : anyType(defaultVal);
}
function eventType() {
  return { type: [Function, Array] };
}
function runEvent(event, ...args) {
  if (typeof event === "function")
    return event(...args);
  else if (Array.isArray(event))
    return event.map((e) => e(...args));
}
async function runAsyncEvent(event, ...args) {
  if (typeof event === "function")
    return await event(...args);
  else if (Array.isArray(event))
    return await Promise.all(event.map((e) => e(...args)));
}
function getEventValue(result, index = 0) {
  if (Array.isArray(result))
    return result[index];
  return result;
}

export {
  stringType,
  numberType,
  booleanType,
  arrayType,
  objectType,
  functionType,
  anyType,
  vNodeType,
  someType,
  eventType,
  runEvent,
  runAsyncEvent,
  getEventValue
};
