// @v-c/utils/dist/get.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/get.ts
var get_exports = {};
__export(get_exports, {
  get: () => get,
  getArr: () => getArr
});
module.exports = __toCommonJS(get_exports);

// src/base.ts
var isString = (val) => typeof val === "string";
var isNumber = (val) => typeof val === "number";
var isBoolean = (val) => typeof val === "boolean";
var isUndefined = (val) => typeof val === "undefined";
var isNull = (val) => val === null;
var isSymbol = (val) => typeof val === "symbol";
var isBaseType = (val) => {
  return isString(val) || isNumber(val) || isBoolean(val) || isSymbol(val) || isUndefined(val) || isNull(val);
};

// src/vnode.ts
var import_vue = require("vue");

// src/lodash.ts
var import_lodash = __toESM(require("lodash.clonedeep"), 1);
var import_lodash2 = __toESM(require("lodash.isequal"), 1);
var import_lodash3 = __toESM(require("lodash.merge"), 1);

// src/lib/hooks/cloned.ts
var import_vue2 = require("vue");

// src/lib/hooks/v-model.ts
var import_vue3 = require("vue");

// src/lib/hooks/state.ts
var import_vue4 = require("vue");

// src/lib/hooks/clipboard.ts
var import_vue5 = require("vue");

// src/lib/hooks/create-injection-state.ts
var import_vue6 = require("vue");

// src/lib/can-use-dom.ts
function canUseDom() {
  return !!(typeof window !== "undefined" && window.document && window.document.createElement);
}
var can_use_dom_default = canUseDom;

// src/lib/hooks/create-global-state.ts
var import_vue7 = require("vue");

// src/lib/hooks/create-shared-composable.ts
var import_vue9 = require("vue");

// src/lib/hooks/try-on-scope-dispose.ts
var import_vue8 = require("vue");

// src/dom/support.ts
var animationEndEventNames = {
  WebkitAnimation: "webkitAnimationEnd",
  OAnimation: "oAnimationEnd",
  animation: "animationend"
};
var transitionEventNames = {
  WebkitTransition: "webkitTransitionEnd",
  OTransition: "oTransitionEnd",
  transition: "transitionend"
};
function supportEnd(names) {
  const el = document.createElement("div");
  for (const name in names) {
    if (names.hasOwnProperty(name) && el.style[name] !== void 0) {
      return {
        end: names[name]
      };
    }
  }
  return false;
}
var animation = can_use_dom_default() && supportEnd(animationEndEventNames);
var transition = can_use_dom_default() && supportEnd(transitionEventNames);

// src/flatten.ts
var import_vue10 = require("vue");

// src/raf.ts
var raf = (callback) => setTimeout(callback, 16);
var caf = (num) => clearTimeout(num);
if (typeof window !== "undefined" && "requestAnimationFrame" in window) {
  raf = (callback) => window.requestAnimationFrame(callback);
  caf = (handle) => window.cancelAnimationFrame(handle);
}
var rafUUID = 0;
var rafIds = /* @__PURE__ */ new Map();
function cleanup(id) {
  rafIds.delete(id);
}
function wrapperRaf(callback, times = 1) {
  rafUUID += 1;
  const id = rafUUID;
  function callRef(leftTimes) {
    if (leftTimes === 0) {
      cleanup(id);
      callback();
    } else {
      const realId = raf(() => {
        callRef(leftTimes - 1);
      });
      rafIds.set(id, realId);
    }
  }
  callRef(times);
  return id;
}
wrapperRaf.cancel = (id) => {
  const realId = rafIds.get(id);
  cleanup(realId);
  return caf(realId);
};

// src/get.ts
var get = (obj, key) => {
  if (!obj)
    return void 0;
  const info = obj[key];
  const keys = key.split(".");
  if (info)
    return info;
  if (keys.length === 1)
    return void 0;
  const popKeys = (0, import_lodash.default)(keys);
  let objData;
  const objKeys = [];
  while (popKeys.length > 0) {
    const keyInfo = popKeys.pop();
    objKeys.unshift(keyInfo);
    const popKey = popKeys.join(".");
    const popInfo = obj[popKey];
    if (popInfo) {
      objData = popInfo;
      break;
    }
  }
  if (!objData)
    return void 0;
  if (objKeys.length === 1)
    return objData[objKeys[0]];
  const allKey = objKeys.join(".");
  const allData = objData[allKey];
  if (allData && isBaseType(allData))
    return allData;
  if (!objData[objKeys[0]])
    return get(objData, objKeys.join("."));
  while (objKeys.length > 1) {
    const firstKey = objKeys.shift();
    const info2 = objData[firstKey];
    if (info2 && isBaseType(info2))
      return info2;
    if (info2 && typeof info2 === "object")
      return get(info2, objKeys.join("."));
    if (objKeys.length === 0)
      break;
  }
  return info;
};
var getArr = (entity, path) => {
  let current = entity;
  for (let i = 0; i < path.length; i += 1) {
    if (current === null || current === void 0)
      return void 0;
    current = current[path[i]];
  }
  return current;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  get,
  getArr
});
