// @v-c/utils/dist/set.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/set.ts
var set_exports = {};
__export(set_exports, {
  default: () => set,
  merge: () => merge2
});
module.exports = __toCommonJS(set_exports);

// src/vnode.ts
var import_vue = require("vue");

// src/lodash.ts
var import_lodash = __toESM(require("lodash.clonedeep"), 1);
var import_lodash2 = __toESM(require("lodash.isequal"), 1);
var import_lodash3 = __toESM(require("lodash.merge"), 1);

// src/lib/hooks/cloned.ts
var import_vue2 = require("vue");

// src/lib/hooks/v-model.ts
var import_vue3 = require("vue");

// src/lib/hooks/state.ts
var import_vue4 = require("vue");

// src/lib/hooks/clipboard.ts
var import_vue5 = require("vue");

// src/lib/hooks/create-injection-state.ts
var import_vue6 = require("vue");

// src/lib/can-use-dom.ts
function canUseDom() {
  return !!(typeof window !== "undefined" && window.document && window.document.createElement);
}
var can_use_dom_default = canUseDom;

// src/lib/hooks/create-global-state.ts
var import_vue7 = require("vue");

// src/lib/hooks/create-shared-composable.ts
var import_vue9 = require("vue");

// src/lib/hooks/try-on-scope-dispose.ts
var import_vue8 = require("vue");

// src/dom/support.ts
var animationEndEventNames = {
  WebkitAnimation: "webkitAnimationEnd",
  OAnimation: "oAnimationEnd",
  animation: "animationend"
};
var transitionEventNames = {
  WebkitTransition: "webkitTransitionEnd",
  OTransition: "oTransitionEnd",
  transition: "transitionend"
};
function supportEnd(names) {
  const el = document.createElement("div");
  for (const name in names) {
    if (names.hasOwnProperty(name) && el.style[name] !== void 0) {
      return {
        end: names[name]
      };
    }
  }
  return false;
}
var animation = can_use_dom_default() && supportEnd(animationEndEventNames);
var transition = can_use_dom_default() && supportEnd(transitionEventNames);

// src/flatten.ts
var import_vue10 = require("vue");

// src/raf.ts
var raf = (callback) => setTimeout(callback, 16);
var caf = (num) => clearTimeout(num);
if (typeof window !== "undefined" && "requestAnimationFrame" in window) {
  raf = (callback) => window.requestAnimationFrame(callback);
  caf = (handle) => window.cancelAnimationFrame(handle);
}
var rafUUID = 0;
var rafIds = /* @__PURE__ */ new Map();
function cleanup(id) {
  rafIds.delete(id);
}
function wrapperRaf(callback, times = 1) {
  rafUUID += 1;
  const id = rafUUID;
  function callRef(leftTimes) {
    if (leftTimes === 0) {
      cleanup(id);
      callback();
    } else {
      const realId = raf(() => {
        callRef(leftTimes - 1);
      });
      rafIds.set(id, realId);
    }
  }
  callRef(times);
  return id;
}
wrapperRaf.cancel = (id) => {
  const realId = rafIds.get(id);
  cleanup(realId);
  return caf(realId);
};

// src/get.ts
var getArr = (entity, path) => {
  let current = entity;
  for (let i = 0; i < path.length; i += 1) {
    if (current === null || current === void 0)
      return void 0;
    current = current[path[i]];
  }
  return current;
};

// src/set.ts
function internalSet(entity, paths, value, removeIfUndefined) {
  if (!paths.length)
    return value;
  const [path, ...restPath] = paths;
  let clone;
  if (!entity && typeof path === "number")
    clone = [];
  else if (Array.isArray(entity))
    clone = [...entity];
  else
    clone = { ...entity };
  if (removeIfUndefined && value === void 0 && restPath.length === 1)
    delete clone[path][restPath[0]];
  else
    clone[path] = internalSet(clone[path], restPath, value, removeIfUndefined);
  return clone;
}
function set(entity, paths, value, removeIfUndefined = false) {
  if (paths.length && removeIfUndefined && value === void 0 && !getArr(entity, paths.slice(0, -1)))
    return entity;
  return internalSet(entity, paths, value, removeIfUndefined);
}
function isObject2(obj) {
  return typeof obj === "object" && obj !== null && Object.getPrototypeOf(obj) === Object.prototype;
}
function createEmpty(source) {
  return Array.isArray(source) ? [] : {};
}
function merge2(...sources) {
  let clone = createEmpty(sources[0]);
  sources.forEach((src) => {
    const loopSet = /* @__PURE__ */ new Set();
    function internalMerge(path) {
      const value = getArr(src, path);
      const isArr = Array.isArray(value);
      if (isArr || isObject2(value)) {
        if (!loopSet.has(value)) {
          loopSet.add(value);
          const originValue = getArr(clone, path);
          if (isArr) {
            clone = set(clone, path, []);
          } else if (!originValue || typeof originValue !== "object") {
            clone = set(clone, path, createEmpty(value));
          }
          Object.keys(value).forEach((key) => {
            internalMerge([...path, key]);
          });
        }
      } else {
        clone = set(clone, path, value);
      }
    }
    internalMerge([]);
  });
  return clone;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  merge
});
