// @v-c/utils/dist/type-util.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/type-util.ts
var type_util_exports = {};
__export(type_util_exports, {
  anyType: () => anyType,
  arrayType: () => arrayType,
  booleanType: () => booleanType,
  eventType: () => eventType,
  functionType: () => functionType,
  getEventValue: () => getEventValue,
  numberType: () => numberType,
  objectType: () => objectType,
  runAsyncEvent: () => runAsyncEvent,
  runEvent: () => runEvent,
  someType: () => someType,
  stringType: () => stringType,
  vNodeType: () => vNodeType
});
module.exports = __toCommonJS(type_util_exports);
function stringType(defaultVal) {
  return {
    type: String,
    default: defaultVal
  };
}
function numberType(defaultVal) {
  return {
    type: Number,
    default: defaultVal
  };
}
function booleanType(defaultVal) {
  return {
    type: Boolean,
    default: defaultVal
  };
}
function arrayType(defaultVal) {
  return {
    type: Array,
    default: defaultVal
  };
}
function objectType(defaultVal) {
  return {
    type: Object,
    default: defaultVal
  };
}
function functionType(defaultVal) {
  return {
    type: Function,
    default: defaultVal
  };
}
function anyType(defaultVal, required) {
  const type = { validator: () => true, default: defaultVal };
  return required ? type : type;
}
function vNodeType() {
  return { validator: () => true };
}
function someType(types, defaultVal) {
  return types ? { type: types, default: defaultVal } : anyType(defaultVal);
}
function eventType() {
  return { type: [Function, Array] };
}
function runEvent(event, ...args) {
  if (typeof event === "function")
    return event(...args);
  else if (Array.isArray(event))
    return event.map((e) => e(...args));
}
async function runAsyncEvent(event, ...args) {
  if (typeof event === "function")
    return await event(...args);
  else if (Array.isArray(event))
    return await Promise.all(event.map((e) => e(...args)));
}
function getEventValue(result, index = 0) {
  if (Array.isArray(result))
    return result[index];
  return result;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  anyType,
  arrayType,
  booleanType,
  eventType,
  functionType,
  getEventValue,
  numberType,
  objectType,
  runAsyncEvent,
  runEvent,
  someType,
  stringType,
  vNodeType
});
