// @v-c/utils/dist/vnode.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/vnode.ts
var vnode_exports = {};
__export(vnode_exports, {
  filterEmpty: () => filterEmpty,
  getSlot: () => getSlot,
  getSlotsProps: () => getSlotsProps,
  hasSlotProps: () => hasSlotProps,
  isEmptyElement: () => isEmptyElement,
  isNodeVShowFalse: () => isNodeVShowFalse
});
module.exports = __toCommonJS(vnode_exports);
var import_vue = require("vue");

// src/base.ts
var isFunction = (val) => typeof val === "function";

// src/vnode.ts
var getSlotsProps = (slots, props, key, ...args) => {
  if (key in slots && isFunction(slots[key])) {
    const slot = slots[key];
    return slot?.(...args);
  }
  if (key in props) {
    const prop = props[key];
    if (isFunction(prop))
      return prop(...args);
    return prop;
  }
  return null;
};
var hasSlotProps = (slots, props, key) => {
  if (key in slots && slots[key])
    return true;
  return !!(key in props && props[key]);
};
function isEmptyElement(c) {
  return c && (c.type === import_vue.Comment || c.type === import_vue.Fragment && c.children.length === 0 || c.type === import_vue.Text && c.children.trim() === "");
}
function filterEmpty(children = []) {
  const res = [];
  children.forEach((child) => {
    if (Array.isArray(child))
      res.push(...child);
    else if (child?.type === import_vue.Fragment)
      res.push(...filterEmpty(child.children));
    else
      res.push(child);
  });
  return res.filter((c) => !isEmptyElement(c));
}
function isNodeVShowFalse(vNode) {
  const showDir = vNode.dirs?.find(({ dir }) => dir === import_vue.vShow);
  return !!(showDir && showDir.value === false);
}
function getSlot(instance, slotName = "default", fallback = []) {
  const slots = instance?.$slots || instance.slots || {};
  const slot = slots[slotName];
  if (slot === void 0)
    return fallback;
  return slot();
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  filterEmpty,
  getSlot,
  getSlotsProps,
  hasSlotProps,
  isEmptyElement,
  isNodeVShowFalse
});
