// @v-c/utils/dist/warning.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/warning.ts
var warning_exports = {};
__export(warning_exports, {
  call: () => call,
  default: () => warning_default,
  note: () => note,
  noteOnce: () => noteOnce,
  resetWarned: () => resetWarned,
  warning: () => warning,
  warningOnce: () => warningOnce
});
module.exports = __toCommonJS(warning_exports);
var warned = {};
function warning(valid, message) {
  if (process.env.NODE_ENV !== "production" && !valid && console !== void 0)
    console.error(`Warning: ${message}`);
}
function note(valid, message) {
  if (process.env.NODE_ENV !== "production" && !valid && console !== void 0)
    console.warn(`Note: ${message}`);
}
function resetWarned() {
  warned = {};
}
function call(method, valid, message) {
  if (!valid && !warned[message]) {
    method(false, message);
    warned[message] = true;
  }
}
function warningOnce(valid, message) {
  call(warning, valid, message);
}
function noteOnce(valid, message) {
  call(note, valid, message);
}
var warning_default = warningOnce;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  call,
  note,
  noteOnce,
  resetWarned,
  warning,
  warningOnce
});
