// unplugin-config/dist/chunk-472H3ZVG.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/core/unplugin.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _fs = require('fs');
var _logger = require('@kirklin/logger');
var _dotenv = require('dotenv'); var _dotenv2 = _interopRequireDefault(_dotenv);

// src/core/constants.ts
var PLUGIN_NAME = "unplugin-config";
var GLOB_CONFIG_FILE_NAME = "_app.config.js";
var OUTPUT_DIR = "dist";
var APP_NAME = PLUGIN_NAME;
var ENV_CONFIG_PREFIX = "VITE_GLOB_";
var OUTPUT_BASE_DIR = "/";

// src/core/utils.ts
var _jsdom = require('jsdom');
var _htmlentities = require('html-entities');
function sanitizeString(str) {
  const sanitized = str.replace(/[^\w.-]/g, "");
  return sanitized.trim().replace(/-/g, "__");
}
function formatPath(path2) {
  return path2.replace(/\\/g, "/");
}
function addScriptToHtmlCode(htmlCode, scriptSrc, position = "head-prepend") {
  const dom = new (0, _jsdom.JSDOM)(htmlCode);
  const scriptElement = dom.window.document.createElement("script");
  scriptElement.src = scriptSrc;
  scriptElement.type = "module";
  const targetElement = position === "body" || position === "body-prepend" ? dom.window.document.body : dom.window.document.head;
  if (position === "head-prepend" || position === "body-prepend") {
    targetElement.insertBefore(scriptElement, targetElement.firstChild);
  } else {
    targetElement.appendChild(scriptElement);
  }
  const updatedHtmlCode = dom.serialize();
  return updatedHtmlCode;
}
function decodeHtmlEntities(encodedString) {
  return _htmlentities.decode.call(void 0, encodedString, { level: "xml", scope: "strict" });
}

// src/core/unplugin.ts
function getAppConfigFileName(options) {
  const shortName = sanitizeString(_optionalChain([options, 'optionalAccess', _ => _.appName]) || APP_NAME);
  return `__PRODUCTION__${shortName}__CONF__`.toUpperCase();
}
function runBuildConfig(options) {
  const config = getEnvConfig(_optionalChain([options, 'optionalAccess', _2 => _2.envVariables, 'optionalAccess', _3 => _3.prefix]), _optionalChain([options, 'optionalAccess', _4 => _4.envVariables, 'optionalAccess', _5 => _5.files]));
  _logger.logger.info(`[${PLUGIN_NAME}] runBuildConfig: ${JSON.stringify(config)}`);
  const configFileName = getAppConfigFileName(options);
  _logger.logger.info(`[${PLUGIN_NAME}] configFileName: ${configFileName}`);
  createConfig({
    configName: configFileName,
    config,
    configFileName: _optionalChain([options, 'optionalAccess', _6 => _6.configFile, 'optionalAccess', _7 => _7.fileName]),
    outputDir: _optionalChain([options, 'optionalAccess', _8 => _8.configFile, 'optionalAccess', _9 => _9.outputDir]) || OUTPUT_DIR,
    appName: _optionalChain([options, 'optionalAccess', _10 => _10.appName]) || APP_NAME
  });
}
function createConfig({ configName, config, configFileName, outputDir, appName }) {
  try {
    const windowConf = `window.${configName}`;
    const configStr = `${windowConf}=${JSON.stringify(config)};
      Object.freeze(${windowConf});
      Object.defineProperty(window, "${configName}", {
        configurable: false,
        writable: false,
      });`.replace(/\n\s*/g, "");
    const outputPath = _path.resolve.call(void 0, _process2.default.cwd(), outputDir);
    _fs.mkdirSync.call(void 0, outputPath, { recursive: true });
    _fs.writeFileSync.call(void 0, 
      _path.resolve.call(void 0, outputPath, configFileName || GLOB_CONFIG_FILE_NAME),
      configStr
    );
    _logger.logger.info(`\u2728 [${appName}] - Configuration file is built successfully:
${outputDir}/${configFileName || GLOB_CONFIG_FILE_NAME}`);
  } catch (error) {
    _logger.logger.error(`[${PLUGIN_NAME}] Configuration file failed to package:
${error instanceof Error ? error.message : String(error)}`);
  }
}
function getEnvConfigFiles() {
  const script = _process2.default.env.npm_lifecycle_script || "";
  const reg = /--mode ([a-z_\d]+)/;
  const mode = _optionalChain([reg, 'access', _11 => _11.exec, 'call', _12 => _12(script), 'optionalAccess', _13 => _13[1]]) || "production";
  return [`.env.${mode}`, ".env"];
}
function getEnvConfig(prefix = ENV_CONFIG_PREFIX, files = getEnvConfigFiles()) {
  let envConfig = {};
  files.forEach((file) => {
    try {
      const env = _dotenv2.default.parse(_fs.readFileSync.call(void 0, _path.resolve.call(void 0, _process2.default.cwd(), file)));
      envConfig = { ...envConfig, ...env };
    } catch (e) {
      console.error(`[${PLUGIN_NAME}] Error in parsing ${file}`, e);
    }
  });
  const reg = new RegExp(`^(${prefix})`);
  Object.keys(envConfig).forEach((key) => {
    if (!reg.test(key)) {
      Reflect.deleteProperty(envConfig, key);
    }
  });
  return envConfig;
}
var unpluginFactory = (options, meta) => {
  const { framework } = meta;
  const ENABLE_PLUGIN = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _14 => _14.configFile, 'optionalAccess', _15 => _15.generate]), () => ( true));
  const generateConfigFile = () => {
    if (ENABLE_PLUGIN) {
      runBuildConfig(options);
    }
    _logger.logger.info(`\u2728 [${_optionalChain([options, 'optionalAccess', _16 => _16.appName]) || APP_NAME}] - Build successfully!`);
  };
  const transformCode = (code) => {
    const { baseDir, configFile, htmlInjection } = options || {};
    const BASE_DIR = baseDir || OUTPUT_BASE_DIR;
    const configFileName = configFile && configFile.fileName || GLOB_CONFIG_FILE_NAME;
    const position = htmlInjection && htmlInjection.position || "head-prepend";
    const shouldDecodeEntities = htmlInjection && htmlInjection.decodeEntities === true;
    const getAppConfigSrc = () => `${BASE_DIR.endsWith("/") ? BASE_DIR : `${BASE_DIR}/`}${configFileName}`;
    let updatedCode = addScriptToHtmlCode(code, getAppConfigSrc(), position);
    if (shouldDecodeEntities) {
      updatedCode = decodeHtmlEntities(updatedCode);
    }
    return updatedCode;
  };
  return {
    name: PLUGIN_NAME,
    vite: {
      apply: "build",
      transformIndexHtml(html) {
        return transformCode(html);
      }
    },
    writeBundle() {
      generateConfigFile();
    },
    transformInclude(id) {
      const formatId = formatPath(id);
      let shouldTransform = false;
      if (ENABLE_PLUGIN) {
        if (!(_nullishCoalesce(_optionalChain([options, 'optionalAccess', _17 => _17.htmlInjection, 'optionalAccess', _18 => _18.enable]), () => ( true)))) {
          return shouldTransform;
        }
        if (formatId.endsWith(".html") && framework !== "webpack" && framework !== "vite") {
          const covertTemplates = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _19 => _19.htmlInjection, 'optionalAccess', _20 => _20.templates, 'optionalAccess', _21 => _21.map, 'call', _22 => _22((template) => formatPath(_path2.default.resolve(_process2.default.cwd(), template)))]), () => ( []));
          if (!_optionalChain([covertTemplates, 'optionalAccess', _23 => _23.length]) || covertTemplates.includes(formatId)) {
            shouldTransform = true;
          }
        }
      }
      return shouldTransform;
    },
    transform(code) {
      return transformCode(code);
    }
  };
};



exports.unpluginFactory = unpluginFactory;
