// unplugin-config/dist/chunk-2FEUO7TK.js
// src/core/unplugin.ts
import process from "node:process";
import path, { resolve } from "node:path";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { logger } from "@kirklin/logger";
import dotenv from "dotenv";

// src/core/constants.ts
var PLUGIN_NAME = "unplugin-config";
var GLOB_CONFIG_FILE_NAME = "_app.config.js";
var OUTPUT_DIR = "dist";
var APP_NAME = PLUGIN_NAME;
var ENV_CONFIG_PREFIX = "VITE_GLOB_";
var OUTPUT_BASE_DIR = "/";

// src/core/utils.ts
import { JSDOM } from "jsdom";
import { decode } from "html-entities";
function sanitizeString(str) {
  const sanitized = str.replace(/[^\w.-]/g, "");
  return sanitized.trim().replace(/-/g, "__");
}
function formatPath(path2) {
  return path2.replace(/\\/g, "/");
}
function addScriptToHtmlCode(htmlCode, scriptSrc, position = "head-prepend") {
  const dom = new JSDOM(htmlCode);
  const scriptElement = dom.window.document.createElement("script");
  scriptElement.src = scriptSrc;
  scriptElement.type = "module";
  const targetElement = position === "body" || position === "body-prepend" ? dom.window.document.body : dom.window.document.head;
  if (position === "head-prepend" || position === "body-prepend") {
    targetElement.insertBefore(scriptElement, targetElement.firstChild);
  } else {
    targetElement.appendChild(scriptElement);
  }
  const updatedHtmlCode = dom.serialize();
  return updatedHtmlCode;
}
function decodeHtmlEntities(encodedString) {
  return decode(encodedString, { level: "xml", scope: "strict" });
}

// src/core/unplugin.ts
function getAppConfigFileName(options) {
  const shortName = sanitizeString(options?.appName || APP_NAME);
  return `__PRODUCTION__${shortName}__CONF__`.toUpperCase();
}
function runBuildConfig(options) {
  const config = getEnvConfig(options?.envVariables?.prefix, options?.envVariables?.files);
  logger.info(`[${PLUGIN_NAME}] runBuildConfig: ${JSON.stringify(config)}`);
  const configFileName = getAppConfigFileName(options);
  logger.info(`[${PLUGIN_NAME}] configFileName: ${configFileName}`);
  createConfig({
    configName: configFileName,
    config,
    configFileName: options?.configFile?.fileName,
    outputDir: options?.configFile?.outputDir || OUTPUT_DIR,
    appName: options?.appName || APP_NAME
  });
}
function createConfig({ configName, config, configFileName, outputDir, appName }) {
  try {
    const windowConf = `window.${configName}`;
    const configStr = `${windowConf}=${JSON.stringify(config)};
      Object.freeze(${windowConf});
      Object.defineProperty(window, "${configName}", {
        configurable: false,
        writable: false,
      });`.replace(/\n\s*/g, "");
    const outputPath = resolve(process.cwd(), outputDir);
    mkdirSync(outputPath, { recursive: true });
    writeFileSync(
      resolve(outputPath, configFileName || GLOB_CONFIG_FILE_NAME),
      configStr
    );
    logger.info(`\u2728 [${appName}] - Configuration file is built successfully:
${outputDir}/${configFileName || GLOB_CONFIG_FILE_NAME}`);
  } catch (error) {
    logger.error(`[${PLUGIN_NAME}] Configuration file failed to package:
${error instanceof Error ? error.message : String(error)}`);
  }
}
function getEnvConfigFiles() {
  const script = process.env.npm_lifecycle_script || "";
  const reg = /--mode ([a-z_\d]+)/;
  const mode = reg.exec(script)?.[1] || "production";
  return [`.env.${mode}`, ".env"];
}
function getEnvConfig(prefix = ENV_CONFIG_PREFIX, files = getEnvConfigFiles()) {
  let envConfig = {};
  files.forEach((file) => {
    try {
      const env = dotenv.parse(readFileSync(resolve(process.cwd(), file)));
      envConfig = { ...envConfig, ...env };
    } catch (e) {
      console.error(`[${PLUGIN_NAME}] Error in parsing ${file}`, e);
    }
  });
  const reg = new RegExp(`^(${prefix})`);
  Object.keys(envConfig).forEach((key) => {
    if (!reg.test(key)) {
      Reflect.deleteProperty(envConfig, key);
    }
  });
  return envConfig;
}
var unpluginFactory = (options, meta) => {
  const { framework } = meta;
  const ENABLE_PLUGIN = options?.configFile?.generate ?? true;
  const generateConfigFile = () => {
    if (ENABLE_PLUGIN) {
      runBuildConfig(options);
    }
    logger.info(`\u2728 [${options?.appName || APP_NAME}] - Build successfully!`);
  };
  const transformCode = (code) => {
    const { baseDir, configFile, htmlInjection } = options || {};
    const BASE_DIR = baseDir || OUTPUT_BASE_DIR;
    const configFileName = configFile && configFile.fileName || GLOB_CONFIG_FILE_NAME;
    const position = htmlInjection && htmlInjection.position || "head-prepend";
    const shouldDecodeEntities = htmlInjection && htmlInjection.decodeEntities === true;
    const getAppConfigSrc = () => `${BASE_DIR.endsWith("/") ? BASE_DIR : `${BASE_DIR}/`}${configFileName}`;
    let updatedCode = addScriptToHtmlCode(code, getAppConfigSrc(), position);
    if (shouldDecodeEntities) {
      updatedCode = decodeHtmlEntities(updatedCode);
    }
    return updatedCode;
  };
  return {
    name: PLUGIN_NAME,
    vite: {
      apply: "build",
      transformIndexHtml(html) {
        return transformCode(html);
      }
    },
    writeBundle() {
      generateConfigFile();
    },
    transformInclude(id) {
      const formatId = formatPath(id);
      let shouldTransform = false;
      if (ENABLE_PLUGIN) {
        if (!(options?.htmlInjection?.enable ?? true)) {
          return shouldTransform;
        }
        if (formatId.endsWith(".html") && framework !== "webpack" && framework !== "vite") {
          const covertTemplates = options?.htmlInjection?.templates?.map((template) => formatPath(path.resolve(process.cwd(), template))) ?? [];
          if (!covertTemplates?.length || covertTemplates.includes(formatId)) {
            shouldTransform = true;
          }
        }
      }
      return shouldTransform;
    },
    transform(code) {
      return transformCode(code);
    }
  };
};

export {
  unpluginFactory
};
