// antdv-style/es/factories/createStyles/index.js
import { computed as p, unref as b } from "vue";
import { useContext as d } from "../../utils/context.js";
import "@emotion/css";
import { createCSS as j } from "../../core/createCSS.js";
import "@emotion/css/create-instance";
import { serializeCSS as E } from "../../core/createSerializeStyles.js";
import { isReactCssResult as F } from "../../utils/css.js";
import { useMediaQueryMap as g, convertResponsiveStyleToString as B } from "./response.js";
function O({ hashPriority: C, useTheme: h, EmotionContext: v }) {
  return (a, r) => {
    const n = r == null ? void 0 : r.__BABEL_FILE_NAME__, m = !!n;
    return (x) => {
      const c = h(), { cache: y } = d(v), { cx: f, css: i } = j(y, {
        hashPriority: (r == null ? void 0 : r.hashPriority) || C,
        label: r == null ? void 0 : r.label
      }), o = g();
      return {
        styles: p(() => {
          let e;
          if (a instanceof Function) {
            const { stylish: t, appearance: s, isDarkMode: l, prefixCls: M, ...S } = c.value, u = (_) => B(_, o.value);
            Object.assign(u, o.value), e = a(
              {
                token: S,
                stylish: t,
                appearance: b(s),
                isDarkMode: b(l),
                prefixCls: M,
                // 工具函数们
                cx: f,
                css: E,
                responsive: u
              },
              x
            );
          } else
            e = a;
          return typeof e == "object" && (F(e) ? e = i(e) : e = Object.fromEntries(
            Object.entries(e).map(([t, s]) => {
              const l = m ? `${n}-${t}` : void 0;
              return typeof s == "object" ? m ? [t, i(s, `label:${l}`)] : [t, i(s)] : [t, s];
            })
          )), e;
        }),
        cx: f,
        theme: c,
        prefixCls: p(() => {
          var e;
          return (e = c.value) == null ? void 0 : e.prefixCls;
        })
      };
    };
  };
}
export {
  O as createStylesFactory
};
