// antdv-style/es/factories/createStyles/response.js
import { computed as m } from "vue";
import "@emotion/css";
import "@emotion/serialize";
import { isReactCssResult as n } from "../../utils/css.js";
import "@emotion/utils";
import "@emotion/css/create-instance";
import { serializeCSS as a } from "../../core/createSerializeStyles.js";
import { useAntdToken as s } from "../../hooks/useAntdToken.js";
import "ant-design-vue";
import { convertBreakpointToResponsive as p } from "../../utils/responsive.js";
function w() {
  const e = s();
  return m(() => {
    const t = {
      xs: `@media (max-width: ${e.value.screenXSMax}px)`,
      sm: `@media (max-width: ${e.value.screenSMMax}px)`,
      md: `@media (max-width: ${e.value.screenMDMax}px)`,
      lg: `@media (max-width: ${e.value.screenLGMax}px)`,
      xl: `@media (max-width: ${e.value.screenXLMax}px)`,
      xxl: `@media (min-width: ${e.value.screenXXLMin}px)`
    };
    return p(t);
  });
}
function S(e, t) {
  return Object.entries(e).map(([i, r]) => {
    let o = r;
    return n(r) || (o = a(r)), t[i] ? `${t[i]} {${o.styles}}` : "";
  }).join("");
}
export {
  S as convertResponsiveStyleToString,
  w as useMediaQueryMap
};
