// antdv-style/es/factories/createThemeProvider/AntdProvider.js
import { defineComponent as M, createVNode as T } from "vue";
import { message as v, notification as P, Modal as y, ConfigProvider as H, theme as r } from "ant-design-vue";
import "@emotion/css";
import "@emotion/serialize";
import "@emotion/utils";
import "@emotion/css/create-instance";
import { useThemeMode as k } from "../../hooks/useThemeMode.js";
const w = /* @__PURE__ */ M((t, {
  slots: a
}) => {
  var o, m, c;
  const [f, l] = v.useMessage((o = t == null ? void 0 : t.staticInstanceConfig) == null ? void 0 : o.message), [d, g] = P.useNotification((m = t == null ? void 0 : t.staticInstanceConfig) == null ? void 0 : m.notification), [h, s] = y.useModal();
  return (c = t == null ? void 0 : t.getStaticInstance) == null || c.call(t, {
    message: f,
    modal: h,
    notification: d
  }), () => {
    const {
      theme: i,
      prefixCls: u
    } = t, {
      appearance: C,
      isDarkMode: I
    } = k(), x = (() => {
      const n = I.value ? r.darkAlgorithm : r.defaultAlgorithm;
      let e = i;
      if (typeof i == "function" && (e = i(C.value)), !e)
        return {
          algorithm: n
        };
      const A = e.algorithm ? Array.isArray(e.algorithm) ? e.algorithm : [e.algorithm] : [];
      return {
        ...e,
        algorithm: e.algorithm ? [n, ...A] : n
      };
    })();
    return T(H, {
      prefixCls: u,
      theme: x
    }, {
      default: () => {
        var n;
        return [l, g, s, (n = a.default) == null ? void 0 : n.call(a)];
      }
    });
  };
}, {
  props: ["theme", "prefixCls", "getStaticInstance", "staticInstanceConfig"]
});
export {
  w as default
};
