// antdv-style/es/factories/createThemeProvider/index.js
import { defineComponent as _, createVNode as o } from "vue";
import { createStyledThemeProvider as i } from "../createStyledThemeProvider.js";
import { DEFAULT_THEME_PROVIDER as F, DEFAULT_THEME_CONTEXT as H } from "../../functions/setupStyled.js";
import { useContext as f } from "../../utils/context.js";
import L from "./AntdProvider.js";
import N from "./ThemeSwitcher.js";
import O from "./TokenContainer.js";
function q(t) {
  const c = t.styledConfig ? i(t.styledConfig) : void 0, {
    StyleEngineContext: m
  } = t;
  return /* @__PURE__ */ _((u, {
    slots: n
  }) => {
    const a = f(m), C = a.CustomThemeContext, s = f(C);
    return () => {
      const {
        customToken: h,
        customStylish: T,
        theme: l,
        getStaticInstance: p,
        prefixCls: x,
        staticInstanceConfig: g,
        appearance: S,
        defaultAppearance: y,
        onAppearanceChange: E,
        themeMode: P,
        defaultThemeMode: v,
        onThemeModeChange: M,
        styled: e
      } = u, {
        prefixCls: A,
        StyledThemeContext: k,
        CustomThemeContext: I
      } = a, D = e ? i(e) : c || F, r = x || A;
      return o(m.Provider, {
        value: {
          prefixCls: r,
          StyledThemeContext: (e == null ? void 0 : e.ThemeContext) || k || H,
          CustomThemeContext: I
        }
      }, {
        default: () => [o(N, {
          themeMode: P,
          defaultThemeMode: v,
          onThemeModeChange: M,
          defaultAppearance: y,
          appearance: S,
          onAppearanceChange: E,
          useTheme: t.useTheme
        }, {
          default: () => [o(L, {
            prefixCls: r,
            staticInstanceConfig: g,
            theme: l,
            getStaticInstance: p
          }, {
            default: () => [o(O, {
              prefixCls: r,
              customToken: h,
              defaultCustomToken: s,
              customStylish: T,
              StyledThemeProvider: D
            }, {
              default: () => {
                var d;
                return [(d = n.default) == null ? void 0 : d.call(n)];
              }
            })]
          })]
        })]
      });
    };
  });
}
export {
  q as createThemeProvider
};
