// antdv-style/es/factories/createThemeProvider/ThemeSwitcher.js
import { defineComponent as k, shallowRef as v, onMounted as P, computed as d, unref as g, createVNode as M, Fragment as C, watchEffect as T, onBeforeUnmount as E } from "vue";
import { useThemeModeProvide as L } from "../../context/ThemeModeContext.js";
import { matchBrowserPrefers as m } from "../../utils/matchBrowserPrefers.js";
import { safeStartTransition as V } from "../../utils/safeStartTransition.js";
import { useMergeValue as w } from "../../utils/useMergeValue.js";
let r;
const b = /* @__PURE__ */ k((e) => {
  const t = () => {
    m("dark").matches ? e.setAppearance("dark") : e.setAppearance("light");
  }, h = () => {
    m("dark").matches ? e.setBrowserPrefers("dark") : e.setBrowserPrefers("light");
  };
  return T(() => {
    if (e.themeMode !== "auto") {
      V(() => {
        e.setAppearance(e.themeMode);
      });
      return;
    }
    setTimeout(t, 1), r || (r = m("dark")), t && r.removeEventListener("change", t), r.addEventListener("change", t);
  }), T(() => {
    r || (r = m("dark")), h && r.removeEventListener("change", h), r.addEventListener("change", h);
  }), E(() => {
    r.removeEventListener("change", t);
  }), () => null;
}, {
  props: ["themeMode", "setAppearance", "setBrowserPrefers"]
}), R = /* @__PURE__ */ k((e, t) => {
  var f;
  const h = e.useTheme(), o = v((f = m("dark")) != null && f.matches ? "dark" : "light"), A = (n) => {
    o.value = n;
  }, u = v();
  P(() => {
    u.value = !0;
  });
  const [c, B] = w("light", {
    value: d(() => e.themeMode),
    defaultValue: (e == null ? void 0 : e.defaultThemeMode) ?? g(h.value.themeMode),
    onChange: (n) => {
      var a;
      return (a = e == null ? void 0 : e.onThemeModeChange) == null ? void 0 : a.call(e, n);
    }
  }), [l, s] = w("light", {
    value: d(() => e.appearance),
    defaultValue: e.defaultAppearance ?? g(h.value.appearance),
    onChange: (n) => {
      var a;
      return (a = e.onAppearanceChange) == null ? void 0 : a.call(e, n);
    }
  });
  return L({
    appearance: l,
    setAppearance: s,
    themeMode: c,
    setThemeMode: B,
    isDarkMode: d(() => l.value === "dark"),
    browserPrefers: o
  }), () => {
    var a, i;
    const n = (i = (a = t == null ? void 0 : t.slots) == null ? void 0 : a.default) == null ? void 0 : i.call(a);
    return M(C, null, [u.value && M(b, {
      themeMode: c.value,
      setAppearance: s,
      setBrowserPrefers: A
    }, null), n]);
  };
}, {
  props: ["appearance", "defaultAppearance", "onAppearanceChange", "themeMode", "defaultThemeMode", "onThemeModeChange", "useTheme"]
});
export {
  R as default
};
