// antdv-style/es/factories/createThemeProvider/TokenContainer.js
import { defineComponent as v, computed as u, reactive as h, createVNode as T } from "vue";
import "@emotion/css";
import "@emotion/serialize";
import "@emotion/utils";
import "@emotion/css/create-instance";
import { serializeCSS as y } from "../../core/createSerializeStyles.js";
import { useThemeMode as C } from "../../hooks/useThemeMode.js";
import "ant-design-vue";
import { useAntdTheme as S } from "../../hooks/useAntdTheme.js";
const b = /* @__PURE__ */ v((e, {
  slots: n
}) => {
  const r = C(), {
    appearance: a,
    isDarkMode: o
  } = r, s = S(), t = u(() => {
    const {
      stylish: i,
      ...m
    } = s.value;
    return {
      stylish: i,
      token: m
    };
  }), l = u(() => e.defaultCustomToken ? e.defaultCustomToken instanceof Function ? e.defaultCustomToken({
    token: t.value.token,
    appearance: a.value,
    isDarkMode: o.value
  }) : e.defaultCustomToken : {}), c = u(() => (e == null ? void 0 : e.customToken) instanceof Function ? {
    ...l.value,
    ...e.customToken({
      token: t.value.token,
      appearance: a.value,
      isDarkMode: o.value
    })
  } : {
    ...l.value,
    ...e == null ? void 0 : e.customToken
  }), k = u(() => e.customStylish ? e.customStylish({
    token: {
      ...t.value.token,
      ...c.value
    },
    stylish: t.value.stylish,
    appearance: a.value,
    isDarkMode: o.value,
    css: y
  }) : {}), f = u(() => ({
    ...k.value,
    ...t.value.stylish
  }));
  return () => {
    const i = {
      ...t.value.token,
      ...c.value,
      stylish: f.value,
      ...h({
        ...r
      }),
      prefixCls: e.prefixCls
    }, m = e.StyledThemeProvider;
    return T(m, {
      theme: i
    }, {
      default: () => {
        var d;
        return [(d = n == null ? void 0 : n.default) == null ? void 0 : d.call(n)];
      }
    });
  };
}, {
  props: ["StyledThemeProvider", "defaultCustomToken", "customToken", "customStylish", "prefixCls"]
});
export {
  b as default
};
