// unocss-preset-chinese/dist/index.mjs
import { toEscapedSelector } from '@unocss/core';

function toArray(value = []) {
  return Array.isArray(value) ? value : [value];
}
function isObject(item) {
  return item && typeof item === "object" && !Array.isArray(item);
}
function mergeDeep(original, patch, mergeArray = false) {
  const o = original;
  const p = patch;
  if (Array.isArray(p)) {
    if (mergeArray && Array.isArray(p)) {
      return [...o, ...p];
    } else {
      return [...p];
    }
  }
  const output = { ...o };
  if (isObject(o) && isObject(p)) {
    Object.keys(p).forEach((key) => {
      if (isObject(o[key]) && isObject(p[key]) || Array.isArray(o[key]) && Array.isArray(p[key])) {
        output[key] = mergeDeep(o[key], p[key], mergeArray);
      } else {
        Object.assign(output, { [key]: p[key] });
      }
    });
  }
  return output;
}
function serializeCSS(cssObj) {
  const cssArr = [];
  for (const prop in cssObj) {
    if (Object.prototype.hasOwnProperty.call(cssObj, prop)) {
      cssArr.push(`${prop}: ${cssObj[prop]};`);
    }
  }
  return cssArr.join(" ");
}

const fontType = ["chinese", "hei", "kai", "song", "imitation-song", "new-song", "li"];

const macosSimplifiedChinese = ["-apple-system", "BlinkMacSystemFont", "PingFang SC", "Hiragino Sans GB", "Heiti SC"];
const macosTraditionalChinese = ["-apple-system", "BlinkMacSystemFont", "PingFang SC"];
const windowsSimplifiedChinese = ["Microsoft YaHei"];
const windowsTraditionalChinese = ["Microsoft Jhenghei"];
const linuxSimplifiedChinese = ["system-ui", "Ubuntu", "Droid Sans", "Source Han Sans SC", "Source Han Sans CN", "Source Han Sans"];
const linuxTraditionalChinese = ["system-ui", "Ubuntu", "Droid Sans", "Source Han Sans TC", "Source Han Sans TW", "Source Han Sans"];
function generateFontList(chineseType, fontType2, fallbackFont = ["sans-serif"], declareEnglishFont = ["Arial"]) {
  const fontList = [];
  if (declareEnglishFont.length > 0) {
    fontList.push(generatePunctuationFontName(fontType2));
    fontList.push(...declareEnglishFont);
  }
  if (chineseType === "simplified") {
    const chineseFonts = [macosSimplifiedChinese, windowsSimplifiedChinese, linuxSimplifiedChinese].flat();
    switch (fontType2) {
      case "chinese":
        fontList.push(...chineseFonts);
        break;
      case "hei":
        fontList.push("PingFang SC", "Heiti SC", "Microsoft YaHei", "Source Han Sans SC", "Source Han Sans CN", "Noto Sans SC", "Noto Sans CJK SC", "WenQuanYi Micro Hei", "WenQuanYi Zen Hei", "SimHei", "WenQuanYi Zen Hei Sharp");
        break;
      case "kai":
        fontList.push("Kaiti SC", "STKaiti", "AR PL UKai CN", "AR PL UKai HK", "AR PL UKai TW", "AR PL UKai TW MBE", "AR PL KaitiM GB", "KaiTi", "KaiTi_GB2312", "DFKai-SB");
        fontList.push(...chineseFonts);
        break;
      case "song":
        fontList.push("Songti SC", "Noto Serif SC", "Noto Serif CJK SC", "Source Han Serif SC", "Source Han Serif CN", "STSong", "AR PL New Sung", "AR PL SungtiL GB", "NSimSun", "SimSun", "WenQuanYi Bitmap Song", "AR PL UMing CN", "PMingLiU", "MingLiU");
        fontList.push(...chineseFonts);
        break;
      case "imitation-song":
        fontList.push("STFangsong", "FangSong", "FangSong_GB2312");
        fontList.push(...chineseFonts);
        break;
      case "new-song":
        fontList.push("SimSun-ExtB", "NSimSun", "Songti SC", "Noto Serif SC", "Noto Serif CJK SC", "Source Han Serif SC", "Source Han Serif CN", "STSong", "AR PL New Sung", "AR PL SungtiL GB", "NSimSun", "SimSun", "WenQuanYi Bitmap Song", "AR PL UMing CN", "PMingLiU", "MingLiU");
        fontList.push(...chineseFonts);
        break;
      case "li":
        fontList.push("STLiti", "LiSu", "Hiragino Mincho ProN W6", "Hiragino Mincho ProN W3");
        fontList.push(...chineseFonts);
        break;
      default:
        fontList.push(...chineseFonts);
    }
  } else {
    const chineseFonts = [macosTraditionalChinese, windowsTraditionalChinese, linuxTraditionalChinese].flat();
    switch (fontType2) {
      case "chinese":
        fontList.push(...chineseFonts);
        break;
      case "hei":
        fontList.push("PingFang SC", "Heiti SC", "Microsoft YaHei", "Source Han Sans SC", "Source Han Sans CN", "Noto Sans SC", "Noto Sans CJK SC", "WenQuanYi Micro Hei", "WenQuanYi Zen Hei", "SimHei", "WenQuanYi Zen Hei Sharp");
        break;
      case "kai":
        fontList.push("Kaiti SC", "STKaiti", "AR PL UKai CN", "AR PL UKai HK", "AR PL UKai TW", "AR PL UKai TW MBE", "AR PL KaitiM GB", "KaiTi", "KaiTi_GB2312", "DFKai-SB", "TW-Kai");
        fontList.push(...chineseFonts);
        break;
      case "song":
        fontList.push("Songti SC", "Noto Serif SC", "Noto Serif CJK SC", "Source Han Serif SC", "Source Han Serif CN", "STSong", "AR PL New Sung", "AR PL SungtiL GB", "NSimSun", "SimSun", "TW-Sung", "WenQuanYi Bitmap Song", "AR PL UMing CN", "AR PL UMing HK", "AR PL UMing TW", "AR PL UMing TW MBE", "PMingLiU", "MingLiU");
        fontList.push(...chineseFonts);
        break;
      case "imitation-song":
        fontList.push("STFangsong", "FangSong", "FangSong_GB2312");
        fontList.push(...chineseFonts);
        break;
      case "new-song":
        fontList.push("SimSun-ExtB", "NSimSun", "Songti SC", "Noto Serif SC", "Noto Serif CJK SC", "Source Han Serif SC", "Source Han Serif CN", "STSong", "AR PL New Sung", "AR PL SungtiL GB", "NSimSun", "SimSun", "WenQuanYi Bitmap Song", "AR PL UMing CN", "PMingLiU", "MingLiU");
        fontList.push(...chineseFonts);
        break;
      case "li":
        fontList.push("LiSu", "STLiti");
        fontList.push(...chineseFonts);
        break;
      default:
        fontList.push(...chineseFonts);
    }
  }
  if (fallbackFont !== null) {
    fontList.push(...fallbackFont);
  }
  return fontList;
}
const defaultChineseFonts = fontType.reduce((fonts, type) => {
  fonts[type] = generateFontList("simplified", type);
  return fonts;
}, {});
function generateFontFaceRule(fontType2, filterFontList) {
  const fontList = defaultChineseFonts[fontType2];
  if (!fontList) {
    throw new Error(`Font type "${fontType2}" is not valid.`);
  }
  const fontName = generatePunctuationFontName(fontType2);
  const filteredFontList = fontList.filter((font) => !filterFontList.includes(font));
  const fontSrc = filteredFontList.map((font) => `local('${font}')`).join(", ");
  return `
@font-face {
  font-family: '${fontName}';
  src: ${fontSrc};
  unicode-range: U+201C, U+201D, U+2018, U+2019, U+2E3A, U+2014, U+2013, U+2026, U+00B7, U+007E, U+002F; 
}
`;
}
function generatePunctuationFontName(fontType2) {
  return `Punctuation ${fontType2.charAt(0).toUpperCase() + fontType2.slice(1)}`;
}

function getNonCjkBlockCss(selectorName, css) {
  return `.${selectorName} em:not(:lang(zh)):not(:lang(ja)):not(:lang(ko)), .${selectorName} em:not(:lang(zh)) { ${isObject(css) ? serializeCSS(css) : css} }`;
}

const getVariables = {
  rootSelector: "var(--un-chinese-typography-root-selector)",
  fontStackSans: "var(--un-chinese-typography-font-stack-sans)",
  fontStackSerif: "var(--un-chinese-typography-font-stack-serif)",
  fontStackMono: "var(--un-chinese-typography-font-stack-mono)",
  fontStackSymbol: "var(--un-chinese-typography-font-stack-symbol)",
  fontFamilyHei: "var(--un-chinese-typography-font-family-hei)",
  fontFamilySong: "var(--un-chinese-typography-font-family-song)",
  fontFamilyKai: "var(--un-chinese-typography-font-family-kai)",
  fontFamilyHeiBlack: "var(--un-chinese-typography-font-family-hei-black)",
  fontFamilySongBlack: "var(--un-chinese-typography-font-family-song-black)",
  fontFamilyKaiBlack: "var(--un-chinese-typography-font-family-kai-black)",
  fontFamilyMono: "var(--un-chinese-typography-font-family-mono)",
  fontWeightBolder: "var(--un-chinese-typography-font-weight-bolder)",
  fontWeightBold: "var(--un-chinese-typography-font-weight-bold)",
  fontWeightNormal: "var(--un-chinese-typography-font-weight-normal)",
  fontWeightLighter: "var(--un-chinese-typography-font-weight-lighter)",
  fontSizeNormal: "var(--un-chinese-typography-font-size-normal)",
  fontSizeXLarge: "var(--un-chinese-typography-font-size-x-large)",
  fontSizeLarge: "var(--un-chinese-typography-font-size-large)",
  fontSizeSmall: "var(--un-chinese-typography-font-size-small)",
  fontSizeXSmall: "var(--un-chinese-typography-font-size-x-small)",
  fontSizeH1: "var(--un-chinese-typography-font-size-h1)",
  fontSizeH2: "var(--un-chinese-typography-font-size-h2)",
  fontSizeH3: "var(--un-chinese-typography-font-size-h3)",
  fontSizeH4: "var(--un-chinese-typography-font-size-h4)",
  fontSizeH5: "var(--un-chinese-typography-font-size-h5)",
  fontSizeH6: "var(--un-chinese-typography-font-size-h6)",
  lineLength: "var(--un-chinese-typography-line-length)",
  lineHeightNormal: "var(--un-chinese-typography-line-height-normal)",
  letterSpacingNormal: "var(--un-chinese-typography-letter-spacing-normal)",
  letterSpacingSmall: "var(--un-chinese-typography-letter-spacing-small)",
  letterSpacingMedium: "var(--un-chinese-typography-letter-spacing-medium)",
  letterSpacingLarge: "var(--un-chinese-typography-letter-spacing-large)",
  lineHeightSizeNormal: "var(--un-chinese-typography-line-height-size-normal)",
  lineHeightSizeLarge: "var(--un-chinese-typography-line-height-size-large)",
  lineHeightSizeXLarge: "var(--un-chinese-typography-line-height-size-x-large)",
  lineHeightSizeSmall: "var(--un-chinese-typography-line-height-size-small)",
  lineHeightSizeXSmall: "var(--un-chinese-typography-line-height-size-x-small)",
  lineHeightSizeH1: "var(--un-chinese-typography-line-height-size-h1)",
  lineHeightSizeH2: "var(--un-chinese-typography-line-height-size-h2)",
  lineHeightSizeH3: "var(--un-chinese-typography-line-height-size-h3)",
  lineHeightSizeH4: "var(--un-chinese-typography-line-height-size-h4)",
  lineHeightSizeH5: "var(--un-chinese-typography-line-height-size-h5)",
  lineHeightSizeH6: "var(--un-chinese-typography-line-height-size-h6)",
  stdBlockUnit: "var(--un-chinese-typography-std-block-unit)",
  stdInlineUnit: "var(--un-chinese-typography-std-inline-unit)",
  textIndentLength: "var(--un-chinese-typography-text-indent-length)",
  textIndentSize: "var(--un-chinese-typography-text-indent-size)",
  chineseQuotePresetsCnHorizontal: "var(--un-chinese-typography-chinese-quote-presets-cn-horizontal)",
  chineseQuotePresetsCnVertical: "var(--un-chinese-typography-chinese-quote-presets-cn-vertical)",
  chineseQuotePresetsTwHorizontal: "var(--un-chinese-typography-chinese-quote-presets-tw-horizontal)",
  chineseQuotePresetsTwVertical: "var(--un-chinese-typography-chinese-quote-presets-tw-vertical)",
  chineseQuotePresetsCommonHorizontal: "var(--un-chinese-typography-chinese-quote-presets-common-horizontal)",
  chineseQuotePresetsCommonVertical: "var(--un-chinese-typography-chinese-quote-presets-common-vertical)",
  chineseQuoteSet: "var(--un-chinese-typography-chinese-quote-set)"
};

function generateBaseCss(selectorName) {
  return `
  
    ${getNonCjkBlockCss(`${selectorName} p`, {
    "text-align": "start"
  })}
    ${getNonCjkBlockCss(selectorName, {
    "letter-spacing": getVariables.letterSpacingNormal
  })}
`;
}

function generateHeadingCss(selectorName) {
  return `
  
    ${getNonCjkBlockCss(`${selectorName} h1, h2, h3`, {
    "letter-spacing": "0"
  })}
    ${getNonCjkBlockCss(selectorName, {
    "letter-spacing": getVariables.letterSpacingNormal
  })}
`;
}

function generateOtherCss(selectorName) {
  return `
.${selectorName} .${selectorName}-meta {
  display: block;
  text-indent: 0;
}
.${selectorName} .${selectorName}-verse {
  text-align: center;
  text-indent: 0;
}
.${selectorName} .${selectorName}-large {
  font-size: ${getVariables.fontSizeLarge};
  line-height: ${getVariables.lineHeightSizeLarge};
}
.${selectorName} .${selectorName}-x-large {
  font-size: ${getVariables.fontSizeXLarge};
  line-height: ${getVariables.lineHeightSizeXLarge};
  letter-spacing: 0.05em;
}
.${selectorName} .${selectorName}-small {
  font-size: ${getVariables.fontSizeSmall};
  line-height: ${getVariables.lineHeightSizeSmall};
}
.${selectorName} .${selectorName}-x-small {
  font-size: ${getVariables.fontSizeXSmall};
  line-height: ${getVariables.lineHeightSizeXSmall};
}
.${selectorName} .${selectorName}-list-latin {
  list-style-type: upper-latin;
}
.${selectorName} .${selectorName}-list-latin ol {
  list-style-type: lower-roman;
}
.${selectorName} .${selectorName}-list-latin ol ol {
  list-style-type: lower-latin;
}
.${selectorName} .${selectorName}-list-han {
  list-style-type: cjk-ideographic;
}
.${selectorName} .${selectorName}-list-han ol {
  list-style-type: decimal;
}
.${selectorName} .${selectorName}-list-han ol ol {
  list-style-type: decimal-leading-zero;
}
.${selectorName} .${selectorName}-fn {
  margin-block-start: 59px;
  border-block-start: 1px solid;
  border-block-start-color: hsl(0deg, 0%, 80%);
  font-size: ${getVariables.fontSizeSmall};
  line-height: ${getVariables.lineHeightSizeNormal};
  font-family: ${getVariables.fontFamilyHei};
}
.${selectorName} .${selectorName}-fn ol {
  margin-block-start: "calc(var(--un-chinese-typography-std-block-unit) * 0.5)";
  margin-block-end: 0;
}
.${selectorName} .${selectorName}-fn li:target {
  background-color: hsl(210deg, 100%, 93%);
}
[data-darkmode=dark] .${selectorName} .${selectorName}-fn li:target {
  background-color: hsl(210deg, 40%, 38%);
}
@media (prefers-color-scheme: dark) {
  [data-darkmode=auto] .${selectorName} .${selectorName}-fn li:target {
    background-color: hsl(210deg, 40%, 38%);
  }
}
.${selectorName} .${selectorName}-hang {
  position: absolute;
  line-height: inherit;
  text-indent: 0;
}
.${selectorName} .${selectorName}-em {
  -webkit-text-emphasis: filled circle;
  -webkit-text-emphasis-position: under;
  text-emphasis: filled circle;
  text-emphasis-position: under right;
}
${getNonCjkBlockCss(`${selectorName}-em`, {
    "-webkit-text-emphasis": "start",
    "text-emphasis": "none"
  })}
.${selectorName} .${selectorName}-ruby--inline {
  display: inline-flex;
  flex-direction: column-reverse;
  height: 1.5em;
  vertical-align: top;
}
.${selectorName} .${selectorName}-ruby--inline rt {
  display: inline;
  margin-bottom: -0.25em;
  line-height: 1;
  text-align: center;
}
.${selectorName} ${selectorName}-spacing {
  display: inline;
}
.${selectorName} ${selectorName}-spacing + sup, .${selectorName} ${selectorName}-spacing + sub {
  margin-inline-start: 0;
}
.${selectorName} .${selectorName}-spacing-start {
  margin-inline-end: 0.25em;
}
.${selectorName} .${selectorName}-spacing-end {
  margin-inline-start: 0.25em;
}
.${selectorName} ${selectorName}-adjacent {
  display: inline;
}
.${selectorName} .${selectorName}-adjacent-half {
  margin-inline-end: -0.5em;
}
.${selectorName} .${selectorName}-adjacent-quarter {
  margin-inline-end: -0.25em;
}
`;
}

const ChineseHeading = {
  "h1, h2, h3, h4, h5, h6": {
    "position": "relative",
    // 顶边距默认为一行间距，且不因边距重叠原因减半
    // 底边距考虑到亲密性，默认为半行间距
    "margin": "0",
    "margin-block-start": "var(--un-chinese-typography-std-block-unit)",
    "margin-block-end": "calc(var(--un-chinese-typography-std-block-unit) * 0.5)",
    "font-weight": "var(--un-chinese-typography-font-weight-bold)",
    "font-size": "var(--un-chinese-typography-font-size-h6)",
    "line-height": "var(--un-chinese-typography-line-height-size-h6)"
  },
  "h1": {
    "margin-block-end": "var(--un-chinese-typography-std-block-unit)",
    "font-size": "var(--un-chinese-typography-font-size-h1)",
    "line-height": "var(--un-chinese-typography-line-height-size-h1)",
    "letter-spacing": "0.05em"
  },
  "h2": {
    "font-size": "var(--un-chinese-typography-font-size-h2)",
    "line-height": "var(--un-chinese-typography-line-height-size-h2)",
    "letter-spacing": "0.05em"
  },
  "h3": {
    "font-size": "var(--un-chinese-typography-font-size-h3)",
    "line-height": "var(--un-chinese-typography-line-height-size-h3)",
    "letter-spacing": "0.05em"
  },
  "h4": {
    "font-size": getVariables.fontSizeH4,
    "line-height": getVariables.lineHeightSizeH4,
    "letter-spacing": "0.05em"
  },
  "h5": {
    "font-size": getVariables.fontSizeH5,
    "line-height": getVariables.lineHeightSizeH5,
    "letter-spacing": "0.05em"
  },
  "h6": {
    "font-size": getVariables.fontSizeH6,
    "line-height": getVariables.lineHeightSizeH6,
    "letter-spacing": "0.05em"
  },
  "h1 + h2, h2 + h3, h3 + h4, h4 + h5, h5 + h6": {
    "margin-block-start": "calc(var(--un-chinese-typography-std-block-unit) * 0.5)"
  }
};

const ChineseTable = {
  "table": {
    "box-sizing": "border-box",
    "table-layout": "fixed",
    "margin-block-start": "calc(var(--un-chinese-typography-std-block-unit) * 0.5)",
    "margin-block-end": "var(--un-chinese-typography-std-block-unit)",
    "margin-inline-start": "auto",
    "margin-inline-end": "auto",
    "border-collapse": "collapse",
    "border-width": "1px",
    "border-style": "solid",
    "border-color": "hsl(0, 0%, 80%)",
    "word-break": "break-word"
  },
  "th, td": {
    "padding-block-start": "calc(var(--un-chinese-typography-std-block-unit) * 0.25)",
    "padding-block-end": "calc(var(--un-chinese-typography-std-block-unit) * 0.25)",
    "padding-inline-start": "calc(var(--un-chinese-typography-std-inline-unit) * 0.5)",
    "padding-inline-end": "calc(var(--un-chinese-typography-std-inline-unit) * 0.5)",
    "border-width": "1px",
    "border-style": "solid",
    "border-color": "hsl(0, 0%, 80%)"
  },
  "caption": {
    "caption-side": "bottom",
    "margin-block-start": "2px",
    "margin-block-end": "-4px",
    "font-size": "var(--un-chinese-typography-font-size-small)",
    "line-height": "var(--un-chinese-typography-line-height-size-small)"
  }
};

const ChineseBase = {
  "blockquote": {
    "margin-block-start": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "margin-block-end": `${getVariables.stdBlockUnit}`,
    "margin-inline-start": `calc(${getVariables.stdInlineUnit} * 2)`,
    "margin-inline-end": `calc(${getVariables.stdInlineUnit} * 2)`,
    "padding-block-start": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "padding-block-end": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "padding-inline-start": getVariables.stdInlineUnit,
    "padding-inline-end": getVariables.stdInlineUnit,
    "background-color": "hsla(0, 0%, 0%, 0.054)"
  },
  "figure": {
    "display": "block",
    "text-align": "center"
  },
  "figure > img": {
    "display": "block",
    "margin-inline-start": "auto",
    "margin-inline-end": "auto"
  },
  "hr": {
    "width": "30%",
    "height": "1px",
    "margin-block-start": `calc(${getVariables.stdBlockUnit} * 2)`,
    "margin-block-end": `calc(${getVariables.stdBlockUnit} * 2 - 1px)`,
    "margin-inline-start": "auto",
    "margin-inline-end": "auto",
    "border": "0",
    "background-color": "hsl(0, 0%, 80%)"
  },
  "p": {
    "margin-block-start": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "margin-block-end": `${getVariables.stdBlockUnit}`,
    "text-align": "justify"
  },
  "pre": {
    "margin-block-start": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "margin-block-end": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "margin-inline-start": "0",
    "margin-inline-end": "0",
    "padding-block-start": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "padding-block-end": `calc(${getVariables.stdBlockUnit} * 0.5)`,
    "padding-inline-start": getVariables.stdInlineUnit,
    "padding-inline-end": getVariables.stdInlineUnit,
    "overflow": "auto",
    "font-family": getVariables.fontFamilyMono,
    "white-space": "pre",
    "word-wrap": "normal",
    "border-radius": "4px",
    "background-color": "hsla(0, 0%, 0%, 0.054)"
  },
  "pre code": {
    "margin": "0",
    "padding": "0",
    "border": "0",
    "border-radius": "0",
    "background-color": "transparent",
    "color": "inherit"
  },
  // Non-CJK letter spacing
  "letter-spacing": getVariables.letterSpacingMedium,
  "a, abbr, code, chinese-spacing, [lang='en-US']": {
    "letter-spacing": "normal"
  }
};

const ChineseInline = {
  "a": {
    "text-decoration": "none"
  },
  "a:hover": {
    "padding-block-end": "1px",
    "border-block-end": "1px solid currentColor",
    "text-decoration": "none"
  },
  "abbr[title]": {
    "padding-block-end": "1px",
    "border-block-end": "1px dotted",
    "text-decoration": "none",
    "cursor": "help"
  },
  "b, strong": {
    "font-weight": "var(--un-chinese-typography-font-weight-bold)"
  },
  "code": {
    "margin-inline-start": "0.25em",
    "margin-inline-end": "0.25em",
    "font-family": "var(--un-chinese-typography-font-family-mono)",
    "font-size": "0.875em"
  },
  "dfn": {
    "font-weight": "var(--un-chinese-typography-font-weight-bold)"
  },
  "em": {
    "font-weight": "var(--un-chinese-typography-font-weight-bold)"
  },
  "figcaption": {
    "display": "inline-block",
    "vertical-align": "top",
    "font-size": "var(--un-chinese-typography-font-size-small)",
    "text-align": "start"
  },
  "i": {
    "font-style": "italic"
  },
  "ins, u": {
    "padding-block-end": "1px",
    "border-block-end": "1px solid",
    "text-decoration": "none"
  },
  "mark": {
    "padding-block-start": "2px",
    "padding-block-end": "2px",
    "padding-inline-start": "1px",
    "padding-inline-end": "1px",
    "margin-inline-start": "1px",
    "margin-inline-end": "1px",
    "background-color": "hsla(58, 100%, 50%, 0.88)",
    "color": "inherit"
  },
  "q": {
    quotes: "var(--un-chinese-typography-chinese-quote-presets-cn-horizontal), var(--un-chinese-typography-chinese-quote-presets-cn-vertical)"
  },
  "rt": {
    "font-size": "0.875em",
    "font-weight": "var(--un-chinese-typography-font-weight-normal)"
  },
  "small": {
    "font-size": "0.875em"
  },
  "strong": {
    "font-weight": "var(--un-chinese-typography-font-weight-bold)"
  },
  "sub, sup": {
    "position": "relative",
    "margin-inline-start": "0.25em",
    "margin-inline-end": "0.25em",
    "font-size": "0.75em",
    "font-family": "var(--un-chinese-typography-font-family-hei)",
    "font-style": "normal",
    "line-height": "1",
    "vertical-align": "baseline"
  },
  "sub": {
    bottom: "-0.25em"
  },
  "sup": {
    top: "-0.5em"
  },
  "sup:target, sup a:target": {
    "background-color": "hsl(210, 100%, 93%)"
  },
  "summary": {
    "padding-inline-start": "1em",
    "outline": "0",
    "cursor": "pointer"
  },
  "summary::-webkit-details-marker": {
    "width": "0.6em",
    "margin-inline-end": "0.4em"
  },
  "u[title]": {
    "cursor": "help",
    "border-block-end-width": "3px",
    "border-block-end-style": "double",
    "border-block-end-color": "hsla(0, 0%, 0%, 0.54)"
  },
  "address, cite, dfn, dt, em": {
    "font-style": "normal"
  },
  "abbr[title], del, ins, s, u": {
    "margin-inline-start": "1px",
    "margin-inline-end": "1px"
  }
};

const ChineseList = {
  "ul, ol, dl": {
    "margin-block-start": "calc(var(--un-chinese-typography-line-height-size-normal) * 0.5)",
    "margin-block-end": "var(--un-chinese-typography-line-height-size-normal)"
  },
  "ul, ol": {
    "padding-inline-start": "var(--un-chinese-typography-text-indent-size)"
  },
  "ul ul, ul ol, ol ul, ol ol": {
    "margin-block-start": "0",
    " margin-block-end": "0"
  },
  "ul": {
    "list-style-type": "disc"
  },
  "ol": {
    "list-style-type": "decimal"
  },
  "li": {
    "list-style-type": "unset"
  },
  "ul ul, ol ul": {
    "list-style-type": "circle"
  },
  "ul ul ul, ul ol ul, ol ul ul, ol ol ul": {
    "list-style-type": "square"
  }
};

const DEFAULT_TYPOGRAPHY = {
  ...ChineseBase,
  ...ChineseHeading,
  ...ChineseList,
  ...ChineseTable,
  ...ChineseInline
};

function getCSS(options) {
  let css = "";
  const { escapedSelector, selectorName, preflights, compatibility } = options;
  const disableNotUtility = compatibility?.noColonNot || compatibility?.noColonWhere;
  for (const selector in preflights) {
    const cssDeclarationBlock = preflights[selector];
    const notChineseSelector = `:not(:where(.not-${selectorName},.not-${selectorName} *))`;
    const pseudoCSSMatchArray = selector.split(",").map((s) => {
      const match = s.match(/:[():\-\w]+$/g);
      if (match) {
        const matchStr = match[0];
        s = s.replace(matchStr, "");
        return escapedSelector.map(
          (e) => disableNotUtility ? `${e} ${s}${matchStr}` : `${e} :where(${s})${notChineseSelector}${matchStr}`
        ).join(",");
      }
      return null;
    }).filter((v) => v);
    if (pseudoCSSMatchArray.length) {
      css += pseudoCSSMatchArray.join(",");
    } else {
      css += escapedSelector.map(
        (e) => disableNotUtility ? selector.split(",").map((s) => `${e} ${s}`).join(",") : `${e} :where(${selector})${notChineseSelector}`
      ).join(",");
    }
    css += "{";
    for (const k in cssDeclarationBlock) {
      const v = cssDeclarationBlock[k];
      css += `${k}:${v};`;
    }
    css += "}";
  }
  return css;
}
function getPreflights(options) {
  const { escapedSelectors, selectorName, cssExtend, compatibility } = options;
  let escapedSelector = Array.from(escapedSelectors);
  if (!escapedSelector[escapedSelector.length - 1].startsWith(".") && !compatibility?.noColonIs) {
    escapedSelector = [`:is(${escapedSelector[escapedSelector.length - 1]},.${selectorName})`];
  }
  if (cssExtend) {
    return getCSS({ escapedSelector, selectorName, preflights: mergeDeep(DEFAULT_TYPOGRAPHY, cssExtend), compatibility });
  }
  return getCSS({ escapedSelector, selectorName, preflights: DEFAULT_TYPOGRAPHY, compatibility });
}

function generateAncientCss(selectorName) {
  return `
.${selectorName}--ancient, .${selectorName}--poetry {
  font-family: ${getVariables.fontFamilySong};
}
.${selectorName}--ancient h1,
.${selectorName}--ancient h2,
.${selectorName}--ancient h3,
.${selectorName}--ancient h4,
.${selectorName}--ancient h5,
.${selectorName}--ancient h6, .${selectorName}--poetry h1,
.${selectorName}--poetry h2,
.${selectorName}--poetry h3,
.${selectorName}--poetry h4,
.${selectorName}--poetry h5,
.${selectorName}--poetry h6 {
  font-family: ${getVariables.fontFamilyKaiBlack};
  font-weight: 800;
  text-align: center;
}
.${selectorName}--ancient h1 .${selectorName}-meta,
.${selectorName}--ancient h2 .${selectorName}-meta,
.${selectorName}--ancient h3 .${selectorName}-meta,
.${selectorName}--ancient h4 .${selectorName}-meta,
.${selectorName}--ancient h5 .${selectorName}-meta,
.${selectorName}--ancient h6 .${selectorName}-meta, .${selectorName}--poetry h1 .${selectorName}-meta,
.${selectorName}--poetry h2 .${selectorName}-meta,
.${selectorName}--poetry h3 .${selectorName}-meta,
.${selectorName}--poetry h4 .${selectorName}-meta,
.${selectorName}--poetry h5 .${selectorName}-meta,
.${selectorName}--poetry h6 .${selectorName}-meta {
  font-weight: 400;
}
@media screen and (min-width: 640px) {
  .${selectorName}--ancient h1 .${selectorName}-meta,
  .${selectorName}--ancient h2 .${selectorName}-meta,
  .${selectorName}--ancient h3 .${selectorName}-meta,
  .${selectorName}--ancient h4 .${selectorName}-meta,
  .${selectorName}--ancient h5 .${selectorName}-meta,
  .${selectorName}--ancient h6 .${selectorName}-meta, .${selectorName}--poetry h1 .${selectorName}-meta,
  .${selectorName}--poetry h2 .${selectorName}-meta,
  .${selectorName}--poetry h3 .${selectorName}-meta,
  .${selectorName}--poetry h4 .${selectorName}-meta,
  .${selectorName}--poetry h5 .${selectorName}-meta,
  .${selectorName}--poetry h6 .${selectorName}-meta {
    position: absolute;
    line-height: inherit;
    text-indent: 0;
    display: inline;
    margin-block-start: 4px;
    margin-inline-start: 8px;
  }
}
.${selectorName}--ancient .${selectorName}-meta, .${selectorName}--poetry .${selectorName}-meta {
  line-height: 24px;
  text-align: center;
  text-indent: 0;
}
.${selectorName}--ancient p {
  text-indent: 2em;
}
.${selectorName}--poetry p {
  text-align: center;
  text-indent: 0;
}
`;
}

function generateDefaultVariables(selectorName, theme) {
  return {
    "--un-chinese-typography-root-selector": `.${selectorName}`,
    "--un-chinese-typography-font-stack-sans": theme?.fontFamily?.sans,
    "--un-chinese-typography-font-stack-serif": theme?.fontFamily?.serif,
    "--un-chinese-typography-font-stack-mono": theme?.fontFamily?.mono,
    "--un-chinese-typography-font-stack-symbol": theme?.fontFamily?.symbol,
    "--un-chinese-typography-font-family-hei": theme?.fontFamily?.hei,
    "--un-chinese-typography-font-family-song": theme?.fontFamily?.song,
    "--un-chinese-typography-font-family-kai": theme?.fontFamily?.kai,
    "--un-chinese-typography-font-family-hei-black": "var(--un-chinese-typography-font-family-hei), font-black",
    "--un-chinese-typography-font-family-song-black": "var(--un-chinese-typography-font-family-song), font-black",
    "--un-chinese-typography-font-family-kai-black": "var(--un-chinese-typography-font-family-kai), font-black",
    "--un-chinese-typography-font-family-mono": "var(--un-chinese-typography-font-stack-mono), monospace, var(--un-chinese-typography-font-stack-symbol)",
    "--un-chinese-typography-font-weight-bolder": "800",
    "--un-chinese-typography-font-weight-bold": "600",
    "--un-chinese-typography-font-weight-normal": "400",
    "--un-chinese-typography-font-weight-lighter": "200",
    "--un-chinese-typography-font-size-normal": "16px",
    "--un-chinese-typography-font-size-x-large": "20px",
    "--un-chinese-typography-font-size-large": "18px",
    "--un-chinese-typography-font-size-small": "14px",
    "--un-chinese-typography-font-size-x-small": "12px",
    "--un-chinese-typography-font-size-h1": "32px",
    "--un-chinese-typography-font-size-h2": "24px",
    "--un-chinese-typography-font-size-h3": "20px",
    "--un-chinese-typography-font-size-h4": "18px",
    "--un-chinese-typography-font-size-h5": "16px",
    "--un-chinese-typography-font-size-h6": "14px",
    "--un-chinese-typography-line-length": "42em",
    "--un-chinese-typography-line-height-normal": "1.5",
    "--un-chinese-typography-letter-spacing-normal": "0",
    "--un-chinese-typography-letter-spacing-small": "0.01em",
    "--un-chinese-typography-letter-spacing-medium": "0.02em",
    "--un-chinese-typography-letter-spacing-large": "0.05em",
    "--un-chinese-typography-line-height-size-normal": "calc(var(--un-chinese-typography-font-size-normal) * var(--un-chinese-typography-line-height-normal))",
    "--un-chinese-typography-line-height-size-large": "var(--un-chinese-typography-line-height-size-normal)",
    "--un-chinese-typography-line-height-size-x-large": "calc(var(--un-chinese-typography-font-size-x-large) * var(--un-chinese-typography-line-height-normal))",
    "--un-chinese-typography-line-height-size-small": "var(--un-chinese-typography-line-height-size-normal)",
    "--un-chinese-typography-line-height-size-x-small": "18px",
    "--un-chinese-typography-line-height-size-h1": "48px",
    "--un-chinese-typography-line-height-size-h2": "36px",
    "--un-chinese-typography-line-height-size-h3": "36px",
    "--un-chinese-typography-line-height-size-h4": "24px",
    "--un-chinese-typography-line-height-size-h5": "24px",
    "--un-chinese-typography-line-height-size-h6": "24px",
    "--un-chinese-typography-std-block-unit": "var(--un-chinese-typography-line-height-size-normal)",
    "--un-chinese-typography-std-inline-unit": "var(--un-chinese-typography-font-size-normal)",
    "--un-chinese-typography-text-indent-length": "2em",
    "--un-chinese-typography-text-indent-size": "calc(var(--un-chinese-typography-font-size-normal) * 2)",
    // 中文引号 Chinese Quote Set
    // `cn`：中华人民共和国国家标准——GB/T 15834-2011《标点符号用法》 http://www.moe.gov.cn/ewebeditor/uploadfile/2015/01/13/20150113091548267.pdf
    // `tw`：中国台湾地区标准——《重訂標點符號手冊》 https://language.moe.gov.tw/001/Upload/FILES/SITE_CONTENT/M0001/HAU/h6.htm
    // `common`：部分中文社区（如知乎）在简体中文里亦采用与中国台湾地区标准一致的规范。
    "--un-chinese-typography-chinese-quote-presets-cn-horizontal": '"\u201C" "\u201D" "\u2018" "\u2019"',
    "--un-chinese-typography-chinese-quote-presets-cn-vertical": '"\u300E" "\u300F" "\u300C" "\u300D"',
    "--un-chinese-typography-chinese-quote-presets-tw-horizontal": '"\u300C" "\u300D" "\u300E" "\u300F"',
    "--un-chinese-typography-chinese-quote-presets-tw-vertical": '"\u300C" "\u300D" "\u300E" "\u300F"',
    "--un-chinese-typography-chinese-quote-presets-common-horizontal": '"\u300C" "\u300D" "\u300E" "\u300F"',
    "--un-chinese-typography-chinese-quote-presets-common-vertical": '"\u300C" "\u300D" "\u300E" "\u300F"',
    "--un-chinese-typography-chinese-quote-set": "common"
  };
}

function generateAnnotationCss(selectorName) {
  return `
     .${selectorName}--annotation p {
      margin-block-start: 0;
      margin-block-end: 0;
      line-height: 2.25;
      text-indent: ${getVariables.textIndentLength};
    }
    .${selectorName}--annotation em {
      -webkit-text-emphasis: filled circle;
      -webkit-text-emphasis-position: under;
      text-emphasis: filled circle;
      text-emphasis-position: under right;
      font-weight: ${getVariables.fontWeightNormal};
    }
    ${getNonCjkBlockCss(`${selectorName}--annotation em`, {
    "-webkit-text-emphasis": "none",
    "text-emphasis": "none"
  })}
    .${selectorName}--annotation .${selectorName}-meta {
        margin-block-start: calc(${getVariables.stdBlockUnit} * 0.5);
        margin-block-end: ${getVariables.stdBlockUnit};
    }
`;
}

function chineseTypography(options) {
  const escapedSelectors = /* @__PURE__ */ new Set();
  const selectorName = options?.selectorName || "chinese";
  const selectorNameRE = new RegExp(`^${selectorName}$`);
  const colorsRE = new RegExp(`^${selectorName}-([-\\w]+)$`);
  const invertRE = new RegExp(`^${selectorName}-invert$`);
  const columnRE = new RegExp(`^${selectorName}--columns-(\\d+)$`);
  const columnPRE = new RegExp(`^${selectorName}--columns-(\\d+) p$`);
  const columnCommaRE = new RegExp(`^${selectorName}--columns-(\\d+) comma$`);
  const columnWidthRE = new RegExp(`^${selectorName}--columns-width-(\\d+)$`);
  const verticalRE = new RegExp(`^${selectorName}--vertical$`);
  const verticalTitleRE = new RegExp(`^${selectorName}--vertical h-(\\d+)$`);
  const cssExtend = options?.cssExtend;
  const compatibility = options?.compatibility;
  return {
    name: "unocss-preset-chinese-typography",
    enforce: "post",
    layers: { typography: -20 },
    rules: [
      [
        selectorNameRE,
        (_, { rawSelector, theme }) => {
          escapedSelectors.add(toEscapedSelector(rawSelector));
          return {
            // 中文每行展示文字（CPL）建议在 30~50 之间，默认 42
            "max-width": getVariables.lineLength,
            // 默认字体大小为 16px，行高 1.5
            "font-size": getVariables.fontSizeNormal,
            "font-weight": getVariables.fontWeightNormal,
            "-webkit-font-smoothing": "subpixel-antialiased",
            "line-height": getVariables.lineHeightNormal,
            // 针对混合英文段落，采取按词折行，长单词通过连词符段行
            // https://justmarkup.com/articles/2015-07-31-dealing-with-long-words-in-css/
            "overflow-wrap": "break-word",
            "word-wrap": "break-word",
            "hyphens": "auto",
            ...generateDefaultVariables(selectorName, theme)
          };
        },
        { layer: "typography" }
      ],
      [
        colorsRE,
        ([, color], { theme }) => {
          const baseColor = theme.colors?.[color];
          if (baseColor == null) {
            return;
          }
          const colorObject = typeof baseColor === "object" ? baseColor : {};
          return {
            "--un-chinese-typography-body": colorObject[700] ?? baseColor,
            "--un-chinese-typography-headings": colorObject[900] ?? baseColor,
            "--un-chinese-typography-links": colorObject[900] ?? baseColor,
            "--un-chinese-typography-lists": colorObject[400] ?? baseColor,
            "--un-chinese-typography-hr": colorObject[200] ?? baseColor,
            "--un-chinese-typography-captions": colorObject[500] ?? baseColor,
            "--un-chinese-typography-code": colorObject[900] ?? baseColor,
            "--un-chinese-typography-borders": colorObject[200] ?? baseColor,
            "--un-chinese-typography-bg-soft": colorObject[100] ?? baseColor,
            // invert colors (dark mode)
            "--un-chinese-typography-invert-body": colorObject[200] ?? baseColor,
            "--un-chinese-typography-invert-headings": colorObject[100] ?? baseColor,
            "--un-chinese-typography-invert-links": colorObject[100] ?? baseColor,
            "--un-chinese-typography-invert-lists": colorObject[500] ?? baseColor,
            "--un-chinese-typography-invert-hr": colorObject[700] ?? baseColor,
            "--un-chinese-typography-invert-captions": colorObject[400] ?? baseColor,
            "--un-chinese-typography-invert-code": colorObject[100] ?? baseColor,
            "--un-chinese-typography-invert-borders": colorObject[700] ?? baseColor,
            "--un-chinese-typography-invert-bg-soft": colorObject[800] ?? baseColor,
            "--un-chinese-typography-font-mono": theme.fontFamily?.mono
          };
        },
        { layer: "typography" }
      ],
      [
        invertRE,
        () => {
          return {
            "--un-chinese-typography-body": "var(--un-chinese-typography-invert-body)",
            "--un-chinese-typography-headings": "var(--un-chinese-typography-invert-headings)",
            "--un-chinese-typography-links": "var(--un-chinese-typography-invert-links)",
            "--un-chinese-typography-lists": "var(--un-chinese-typography-invert-lists)",
            "--un-chinese-typography-hr": "var(--un-chinese-typography-invert-hr)",
            "--un-chinese-typography-captions": "var(--un-chinese-typography-invert-captions)",
            "--un-chinese-typography-code": "var(--un-chinese-typography-invert-code)",
            "--un-chinese-typography-borders": "var(--un-chinese-typography-invert-borders)",
            "--un-chinese-typography-bg-soft": "var(--un-chinese-typography-invert-bg-soft)"
          };
        },
        { layer: "typography" }
      ],
      [
        columnRE,
        ([, d]) => ({ "column-count": Number(d) }),
        { layer: "typography" }
      ],
      [
        columnCommaRE,
        () => ({ "max-width": "none", "column-gap": "2em" }),
        { layer: "typography" }
      ],
      [
        columnPRE,
        () => ({
          "margin-block-start": `calc(${getVariables.stdBlockUnit} * 0.5 * 0.5 )`,
          "margin-block-end": `calc(${getVariables.stdBlockUnit} * 0.5 )`,
          "text-indent": `calc(${getVariables.textIndentLength} * 0.5 * 0.5 )`
        }),
        { layer: "typography" }
      ],
      [
        columnWidthRE,
        ([, d]) => ({ "column-width": `${Number(d)}em` }),
        { layer: "typography" }
      ],
      [
        verticalRE,
        () => {
          return {
            "max-width": "none",
            "max-height": "var(--un-chinese-typography-line-length)",
            "writing-mode": "vertical-rl",
            "letter-spacing": "0.125em"
          };
        },
        { layer: "typography" }
      ],
      [
        verticalTitleRE,
        () => ({ "text-align": "start" }),
        { layer: "typography" }
      ]
    ],
    preflights: [
      {
        layer: "typography",
        getCSS: () => {
          return generateBaseCss(selectorName);
        }
      },
      {
        layer: "typography",
        getCSS: () => {
          return generateHeadingCss(selectorName);
        }
      },
      {
        layer: "typography",
        getCSS: () => {
          return generateOtherCss(selectorName);
        }
      },
      {
        layer: "typography",
        getCSS: () => {
          return generateAncientCss(selectorName);
        }
      },
      {
        layer: "typography",
        getCSS: () => {
          return generateAnnotationCss(selectorName);
        }
      },
      {
        layer: "typography",
        getCSS: () => {
          if (escapedSelectors.size > 0) {
            return getPreflights({ escapedSelectors, selectorName, cssExtend, compatibility });
          }
        }
      }
    ]
  };
}

function presetChinese(options = {}) {
  const {
    extendTheme = true,
    themeKey = "fontFamily",
    chineseType = "simplified",
    fallbackFont = ["sans-serif"],
    declareEnglishFont = ["Arial"]
  } = options;
  const fonts = {
    ...fontType.reduce((fonts2, type) => {
      fonts2[type] = generateFontList(chineseType, type, fallbackFont, declareEnglishFont);
      return fonts2;
    }, {}),
    ...options?.fonts
  };
  const fontObject = Object.fromEntries(
    Object.entries(fonts || {}).map(([name, meta]) => [name, toArray(meta).map((m) => m)])
  );
  const preset = {
    name: "unocss-preset-chinese"
  };
  if (extendTheme) {
    preset.extendTheme = (theme) => {
      if (!theme[themeKey]) {
        theme[themeKey] = {};
      }
      for (const [name, fonts2] of Object.entries(fontObject)) {
        const fontString = fonts2.join(",");
        if (typeof theme[themeKey][name] === "string") {
          theme[themeKey][name] = `${fontString},${theme[themeKey][name]}`;
        } else {
          theme[themeKey][name] = fontString;
        }
      }
    };
  }
  const fontFacePreflights = Object.keys(defaultChineseFonts).map((fontType2) => ({
    getCSS: () => generateFontFaceRule(fontType2, declareEnglishFont)
  }));
  preset.preflights = fontFacePreflights;
  return preset;
}

export { chineseTypography, presetChinese as default };
