// antdv-style/es/functions/createInstance.js
import { CacheManager as p } from "../core/CacheManager.js";
import { createCSS as x } from "../core/createCSS.js";
import T from "@emotion/css/create-instance";
import { serializeCSS as S } from "../core/createSerializeStyles.js";
import { createEmotionContext as _ } from "../factories/createEmotionContext.js";
import { createContext as s } from "../utils/context.js";
import { createUseTheme as g } from "../factories/createUseTheme.js";
import { createStylesFactory as k } from "../factories/createStyles/index.js";
import { createThemeProvider as u } from "../factories/createThemeProvider/index.js";
const a = new p();
typeof global < "u" && (global.__ANTD_STYLE_CACHE_MANAGER_FOR_SSR__ = a);
function G(o) {
  var n;
  const e = {
    ...o,
    key: o.key ?? "zcss",
    // 新建的 instance key 如果不传，则设为 zcss- 使得该 key 和 acss 不一样
    speedy: o.speedy ?? !1
  }, t = T({
    key: e.key,
    speedy: e.speedy,
    container: e.container
  }), m = _(t);
  t.cache = a.add(t.cache);
  const i = s(
    e.customToken ? e.customToken : {}
  ), y = (n = e.styled) == null ? void 0 : n.ThemeContext, c = s({
    CustomThemeContext: i,
    StyledThemeContext: y,
    prefixCls: e == null ? void 0 : e.prefixCls
  }), r = g({ StyleEngineContext: c }), h = k({
    hashPriority: e.hashPriority,
    useTheme: r,
    EmotionContext: m
  }), f = u({
    styledConfig: e.styled,
    StyleEngineContext: c,
    useTheme: r
  }), { cx: l } = x(t.cache, { hashPriority: e.hashPriority }), { injectGlobal: C, keyframes: d } = t;
  return {
    // ******************** //
    // **** 样式生成相关 **** //
    // ******************** //
    createStyles: h,
    //* ******************* //
    //* ***  样式表管理  **** //
    //* ******************* //
    styleManager: t,
    // ******************** //
    // **** 基础样式方法 **** //
    // ******************** //
    css: S,
    cx: l,
    keyframes: d,
    injectGlobal: C,
    // ******************** //
    // ***** 主题相关 ***** //
    // ******************** //
    useTheme: r,
    ThemeProvider: f
  };
}
export {
  G as createInstance
};
